Describing steps to support out-of-tree providers (#463)
openshift/enhancements: enhancements/machine-api/out-of-tree-provider-support.md

---

> title: neat-enhancement-idea

Reviewer: will need to update this title

authors:
  - "@Danil-Grigorev"
reviewers:
  - "@enxebre"
  - "@elmiko"
  - "@mfedosin"
approvers:
  - "@enxebre"
  - "@elmiko"
  - TBD
creation-date: 2020-08-31
last-updated: 2020-09-01

> status: implementable

Reviewer: Given Draft / WIP status, provisional is probably more appropriate?

Author: It seems it is closer to the implementable phase now :)

see-also:
  - "/enhancements/cloud-controller-manager/openstack-cloud-controller-manager.md"  
replaces:
superseded-by:
---

# Add support for out-of-tree cloud provider integration

## Release Signoff Checklist

- [x] Enhancement is `implementable`
- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

## Open Questions

1) Should we running `cloud-controller-manager` as:
- static `Pod` managed from the `Node` filesystem.
- `Deployment` backing up the `Pod`.
- `ReplicaSet` to run a `Pod` on each master node.

## Summary

> This enhancement proposal describes steps needed to add support for out-of-tree providers in cloud providers currently supported by Openshift. This includes: AWS, Azure, GCP, vSphere, IBM. Openstack provider proposal is located at it's dedicated document.

Reviewer: Indeed, because doesn't the OpenStack document propose that they will implement this as part of supporting the OpenStack provider?  Is this enhancement meant to provide the details needed?

Right now it looks like the two enhancements should be merged into one.  #423

> This enhancement proposal describes steps needed to add support for out-of-tree providers in cloud providers currently supported by Openshift. This includes: AWS, Azure, GCP, vSphere, IBM. Openstack provider proposal is located at it's dedicated document.

Author: We merged these two together here :+1:

## Motivation

Upstream is currently moving towards exclusion of in-tree cloud-provider code from `k/k` repository, and instead introducing out-of-tree provider support, making them independent from `k/k` releases, and simplifying the process of development and e2e testing for those plugins. Previously each cloud-provider was located in the [`legacy-cloud-providers`](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/legacy-cloud-providers) and now the support and main development is done in the dedicated cloud-provider [repositories](https://github.com/kubernetes?q=cloud-provider-). Upstream is planning to remove in-tree support completely in the future releases, approximetly in [v1.21](https://kubernetes.slack.com/archives/CPPU7NY8Y/p1589289501020600).

### Goals

- Prepare openshift components to accomodate the out-of-tree plugins and remove support for in-tree implementation.
- Add an installer option to select an`external` cloud provider.

> - Deploy cluster infrastructure accordingly to selected `external` option for cloud provider, and allow `cloud-controller` component registration via the provider specific upstream implemetation.

Author: There is a need for a change in the `Infrastructure` resource. We also need a conditional interpretation of the `Infrastructure` resource from the same `library-go` implementation. In the first scenario, it recognizes the `external` option as a request to ignore the `cloud-config` and the actual `platform` value, and in the second it should pick both real values and ignore the external option selection. I'll add a chapter describing this in more details.

### Non-Goals

> - Force immediate exclusion of in-tree support for currently used providers, as their out-of-tree counterparts are added.

Reviewer: Does this imply there would be some period where OpenShift would support both?  That doesn't seem ideal ...

As @JoelSpeed noted in another comment, "Surely a user doesn't need to know about this implementation detail?"

> - Force immediate exclusion of in-tree support for currently used providers, as their out-of-tree counterparts are added.

Author: If it is done correctly, we could end up with a situation when one provider will be using external configuration, and a couple of others will still default to the in-tree approach. For the provider ready to be provisioned with the out-of-tree `CCM`, we could upgrade to this configuration and no longer support in-tree since the transition.

That's how I see this. Correct me if I'm missing something, but essentially if this ideally should be a painless approach.

> - Force immediate exclusion of in-tree support for currently used providers, as their out-of-tree counterparts are added.

Reviewer 2: Can you clarify why Is this a non-goal? I think what we need is section on how we'd handle downgrades.

> - Force immediate exclusion of in-tree support for currently used providers, as their out-of-tree counterparts are added.

Author: Yes, starting with OpenStack in 4.7 and following up with other providers is in the plan. I can see no reason to move out to out-of-tree for AWS tomorrow, as the code there is just a straight import from the in-tree implementation - https://github.com/kubernetes/cloud-provider-aws/blob/master/cmd/aws-cloud-controller-manager/main.go#L57-L61

> - Force immediate exclusion of in-tree support for currently used providers, as their out-of-tree counterparts are added.

Reviewer 3: Perhaps this can be reworded to make it more obvious what we mean by this? Does this suggestion capture the correct intent?
```suggestion
- Force an immediate transition for all cloud providers from in-tree support to their out-of-tree counterparts.
```

> - Force immediate exclusion of in-tree support for currently used providers, as their out-of-tree counterparts are added.

Reviewer 4: Note that we need to validate each internal-to-external migration to make sure it's safe. eg, the last time I looked, the Azure out-of-tree provider was lagging several months behind the in-tree provider, so we'd regress people if we switched Azure to out-of-tree.

- Add support for CSI and other out-of-tree implementations of cluster storage interfaces.
- Add support for running multiple cloud configurations in a single cluster.

## Proposal

> Add an out-of-tree support based on current IPI, and make a seamless transition for all currently supported providers.

Author: Right, going to remove it, as it is an overall infrastructure change affecting both UPI and IPI.

> This change will be gradually executed on all suported cloud providers: AWS, Azure, GCP, vSphere and IBM.

Reviewer 2: I agree. Also worth mention that any possibly upcoming provider will follow same approach, e.g oVirt.

> This change will be gradually executed on all suported cloud providers: AWS, Azure, GCP, vSphere and IBM.

Author: SGTM. We decided to merge the Openstack proposal into examples in this enhancement with Michael.

### User Stories

> As a cloud developer, I’d like to improve support for cloud features for openshift

Author: I was thinking of that as a prefix to each user story :)

#### Story 1:

We’d like to deploy control-plane node on a spot instance, to allowing customers to get additional cost reduction on their clusters. In order to introduce and test this feature, we have to implement this in the kubernetes repository, first, then wait for the next release to get a “free” rebase on the feature. We are also bound to the kubernetes release cycles, making it hard to achieve the goal in case the upstream is already going through the feature freeze. Starting downstream is not an option, as the implementation could be rejected upstream or gain less attention then needs to quickly get it into the release, diverging our fork from the source.

#### Story 2:

> As an openshift developer I’m assigned a release blocker BZ regarding upstream implementation. Proving the value for merging this and making the bug a release blocker upstream requires careful communication with upstream maintainers, and explaining the value to the people whom are not necessarily involved into internal details, so their opinion could differ and delay merging an important fix.

Reviewer 3: I get what you're going for here, but you don't really explain how this proposal is helping in this scenario. I think the value here is we would have better control over the codebase within the team reducing external blockers?

> As an openshift developer responsible for the cloud controllers, I want to have control of the cloud provider code independently from the core of the product so that my team can deliver important fixes in a timely manner without being blocked by the wider community for extended periods.

Reviewer 5: No matter if in- or out-of-tree provider I'd expect a BZ fix land upstream.

Reviewer 5: You can test that in the openshift cluster, but the end goal is we don't diverge more than we need to, and per diverge I mean here we backport upstream accepted fixes to previous releases where we find it required.

Author: Added this story. Thanks :)

#### Story 3:

We desire to extend our team responsibilities on upstream cloud providers in the future and gain weight in promoting features into upstream. Getting such for kubernetes repository is harder for us and maintainers, as this would mean giving weight for approving and merging features for the parts of the kubernetes project, not necessarily laying under our responsibilities and scope of knowledge.

#### Story 4:

> We’d like to discuss technical details related to a specific cloud in a SIG meeting with people who are also involved into development in this domain, and that way gain useful insights into the cloud infrastructure, improve the overall quality of our features, stay on top of the new features, and improve the relations with maintainers outside of our company, which nevertheless share with us a common goal.

Reviewer 3: These are valuable comments for the motivation behind this project, though I'm not sure they fit the user story definition. Typically a user story follows a format similar to:

> As a < type of user >, I want < some goal > so that < some reason >.

Reviewer 2: I agree. I'd suggest we rephrase the stories to stick to the conventional format.
I think the most relevant is https://github.com/openshift/enhancements/pull/463/files#r490094319.
May be also something in the line of "As an OCP maintainer I want to make sure we accommodate the platform to build atop the latest kubernetes release."

### Implementation Details/Notes/Constraints

> The `external CCM` implementation will be hosted within openshift repository in https://github.com/cluster-cloud-controller-manager-operator. That repository will manage `CCM` provisioning for metioned cloud providers + `Openstack` out-of-tree provider implementation.

Reviewer 3: Tip, you haven't used the `CCM` acronym before this point in the document, when you introduce a ancronym it's best to put it beside the fully expanded version for the first time, so for example:

> A Three Letter Acronym (TLA) can be used to provide a short reference to some subject. TLAs are typically used by for the sake of brevity when a longer subject is discussed in detail.

> The `external CCM` implementation will be hosted within openshift repository in https://github.com/cluster-cloud-controller-manager-operator. That repository will manage `CCM` provisioning for metioned cloud providers + `Openstack` out-of-tree provider implementation.

Reviewer: missing `openshift` in the git repo URL

> `Cloud-controller` will be provisioned in the `openshift-cloud-controller` namespace.

Reviewer 3: Should this be `cloud-controller-manager`?

- Deploy a `cloud-controller` pod in the `openshift-cloud-controller` namespace.

## Design Details

### Kubelet

Openshift manages `kubelet` configuration with `machine-config-operator`. The actual `cloud-config` arguments are passed externally from the `Infrastructure` resource on https://github.com/openshift/machine-config-operator/blob/2a51e49b0835bbd3ac60baa685299e86777b064f/pkg/controller/template/render.go#L310-L357, where the support for `external` cloud provider should be added. For each platform with `external` configuration, the `cloud-config` flag should not be set.

### Kube-api-server

No changes for the current configuration are needed

### Kube-controller-manager

TBD

#### Upgrade/Downgrade strategy

Upstream [proposal](https://github.com/kubernetes/enhancements/blob/473d6a094f07e399079d1ce4698d6e52ddcc1567/keps/sig-cloud-provider/20190422-cloud-controller-manager-migration.md#motivation) describes main pinpoints for migration between in-tree and out-of-tree for HA clusters. They propose implementation for cross-namespace leadership delegation, which would help Openshift to preserve HA during upgrages.

> In case the cluster does not require to be HA, then upgrading under CVO management, which will lead to two simultanious replicas of `cloud-controllers` running at the same time at some moment is not a problem. Each provider, therefore supports multiple options to run the `cloud-controller`:

Reviewer 3: > running at the same time at some moment is not a problem.

This part of this statement should be justified.

> When two replicas of the `cloud-controllers` are running simultaneously, <something happens and that's ok because...>

> 1) As a `DaemonSet`. This is the HA option, where the leader-election on the running pod ensuring a single replica is serving cluster at any point in time.

Reviewer 3: I think how we plan to run the CCM should be higher up in the proposal, in the core of it, if we are running as a DaemonSet that should be explained.

Author: Picking up over deployment there, the explanation (or just a decision)  will be in place.

#### Examples

##### AWS

Main repository: https://github.com/kubernetes/cloud-provider-aws/
Sample manifests: https://github.com/kubernetes/cloud-provider-aws/tree/master/manifests

To try external cloud-controller-manager, siply run:
```bash
oc apply -f https://raw.githubusercontent.com/kubernetes/cloud-provider-aws/master/manifests/rbac.yaml  
oc apply -f https://raw.githubusercontent.com/kubernetes/cloud-provider-aws/master/manifests/aws-cloud-controller-manager-daemonset.yaml
```

##### Azure

Main repository: https://github.com/kubernetes-sigs/cloud-provider-azure
Sample manifests: https://github.com/kubernetes-sigs/cloud-provider-azure/tree/master/examples/out-of-tree

##### GCP

Main repository: https://github.com/kubernetes/cloud-provider-gcp
Sample manifests: https://github.com/kubernetes/cloud-provider-gcp/tree/master/deploy

##### vSphere

Main repository: https://github.com/kubernetes/cloud-provider-vsphere
Sample manifests: https://github.com/kubernetes/cloud-provider-vsphere/tree/master/manifests/controller-manager

### Test Plan

> - Each provider at the time of migration should have a working CI implemetation, which will assist in testing the provisioning with out-of-tree support enabled.

Reviewer 3: Do you think existing CI should be sufficient to cover this, or do we need to add new things? (The answer to this should be reflected into the doc)

Author: I would assume that having CI tests to execute upgrades on each provider onto cluster configuration managed externally, will be enough to confirm this is working. We could gain this for free on the final PR implementing the default `external` option in the installer, but probably the better way (or an additional milestone) will be to replicate these steps in CI for the PRs published in the proposed https://github.com/openshift/cluster-cloud-controller-manager-operator repository.

Author: Proposed annotation on the `Infrastructure` resource might help with CI here too.

- Ensure that transition between in-tree and out-of-tree will be handled by the `Infrastracture` `external-cloud-config` field being set.
- Ensure in-tree and out-of-tree providers could co-exist in the cluster if this is supported.
- Ensure that upgrade from a release running on in-tree provider only to the out-of-tree succeeds.

##### Removing a deprecated feature

> Upstream is currently planning to remove support for in-tree providers with OCP-4.8 (kubernetes 1.21) - based on the slack [conversation](https://kubernetes.slack.com/archives/CPPU7NY8Y/p1589289501020600). The enchancement [describes](https://github.com/kubernetes/enhancements/blob/473d6a094f07e399079d1ce4698d6e52ddcc1567/keps/sig-cloud-provider/20190125-removing-in-tree-providers.md) steps towards this process. Our goal is to have a working alternative for in-tree providers ready to be used on-par with old implementation befor the upstream release will remove this feature from main `k/k` repository.

Reviewer 3: ```suggestion
Upstream is currently planning to remove support for in-tree providers with OCP-4.8 (kubernetes 1.21) - based on the slack [conversation](https://kubernetes.slack.com/archives/CPPU7NY8Y/p1589289501020600). The enchancement [describes](https://github.com/kubernetes/enhancements/blob/473d6a094f07e399079d1ce4698d6e52ddcc1567/keps/sig-cloud-provider/20190125-removing-in-tree-providers.md) steps towards this process. Our goal is to have a working alternative for in-tree providers ready to be used on-par with old implementation before the upstream release will remove this feature from the core Kubernetes repository.
```

## Timeline

Follow upstream discussions and implementation/releases

## Kubernetes 1.18

- vSphere support graduaded to beta: https://github.com/kubernetes/enhancements/issues/670

## Kubernetes 1.19

- Azure support goes into beta: https://github.com/kubernetes/enhancements/issues/667

## Infrastructure Needed

> (Possibly) Forks for out-of-tree provider repositories:

Reviewer 3: Remove (Possibly), to enable us to do bug fixes in an efficient manner, we will need to have a fork of each of these projects so that we can control what we ship and port to upstream later

---
title: out-of-tree-provider-support
authors:
  - "@Danil-Grigorev"
reviewers:
  - "@enxebre"
  - "@elmiko"
  - "@mfedosin"
approvers:
  - "@enxebre"
  - "@elmiko"
  - TBD
creation-date: 2020-08-31
last-updated: 2020-09-12
status: implementable
see-also:
  - "/enhancements/cloud-controller-manager/openstack-cloud-controller-manager.md"  
replaces:
superseded-by:
---

# Add support for out-of-tree cloud provider integration

## Release Signoff Checklist

- [x] Enhancement is `implementable`
- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

## Summary

This enhancement proposal describes steps needed to add support for out-of-tree providers in cloud providers currently supported by Openshift. This includes: AWS, Azure, GCP, vSphere, IBM. Openstack provider proposal is located at it's dedicated [document](https://github.com/openshift/enhancements/tree/master/enhancements/cloud-controller-manager/openstack-cloud-controller-manager.md).

## Motivation

> The Kubernetes community is currently moving towards exclusion of the in-tree cloud-provider code from the core Kubernetes repository, and instead introducing out-of-tree provider support, making them independent from the core Kubernetes release cycle, and simplifying the process of development and e2e testing for those plugins. Previously each cloud-provider was located in the [`legacy-cloud-providers`](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/legacy-cloud-providers) and now the support and main development is done in the dedicated cloud-provider [repositories](https://github.com/kubernetes?q=cloud-provider-). They plan to remove the in-tree support completely in the future releases, approximetly in [v1.21](https://kubernetes.slack.com/archives/CPPU7NY8Y/p1589289501020600).

Reviewer 6: is there one out-of-tree binary, or several with different release cycles each?

Author: Different, with different release cycles. Will poke upstream for the readiness for GA about those. Currently, there is some work being done towards implementing `InstancesV2` in AWS.

### Goals

- Prepare openshift components to accomodate the out-of-tree plugins and remove support for the in-tree implementation.
- Describe an approach to select an `external` cloud provider installation for the cloud.

> - Provide the means to allow generic `cloud-controller` component registration via the provider specific upstream implemetation.

Reviewer 2: Probably want to include a goal to make sure we don't miss feature parity with what we have today in terms of cloud integrations during the transition.

### Non-Goals

- Force immediate exclusion of in-tree support for currently used providers, as their out-of-tree counterparts are added.
- Add support for CSI and other out-of-tree implementations of cluster storage interfaces.
- Add support for running multiple cloud configurations in a single cluster.

## Proposal

> Add an out-of-tree support based on current IPI with minor adjustments, and make a seamless transition for all currently supported providers.

Reviewer 2: why is this tied to IPI? Transition through out of tree needs to be driven via OCP payload release regardless how the cluster was installed. We probably need a section to elaborate UPI caveats.

This change will be gradually executed on all suported cloud providers: AWS, Azure, GCP, vSphere and IBM.

### User Stories

As a cloud developer, I’d like to improve support for cloud features for openshift

#### Story 1:

> We’d like to deploy control-plane node on a spot instance, to allowing customers to get additional cost reduction on their clusters. In order to introduce and test this feature, we have to implement this in the kubernetes repository, first, then wait for the next release to get a “free” rebase on the feature. We are also bound to the kubernetes release cycles, making it hard to achieve the goal in case the core Kubernetes is already going through the feature freeze. Starting downstream is not an option, as the implementation could be rejected upstream or gain less attention then needs to quickly get it into the release, diverging our fork from the source.

Reviewer 7: No, we don't want to support this or have to support this.  No way.

Reviewer 2: I agree this is not the best example. Can we please rephrase this to something like "As a provider owner I want to develop, built, and release fixes and features independently of Kubernetes core"

#### Story 2:

As an openshift developer I’m assigned a release blocker BZ regarding upstream implementation. Proving the value for merging this and making the bug a release blocker upstream requires careful communication with upstream maintainers, and explaining the value to the people whom are not necessarily involved into internal details, so their opinion could differ and delay merging an important fix.

#### Story 3:

> We desire to extend our team responsibilities on upstream cloud providers in the future and gain weight in promoting features into upstream. Getting such for kubernetes repository is harder for us and maintainers, as this would mean giving weight for approving and merging features for the parts of the kubernetes project, not necessarily laying under our responsibilities and scope of knowledge.

Reviewer 5: This is not correct. Even already each sig/community member can work and does work only on a specific area of code and can easily be reviewer, approver and owner of that code.

#### Story 4:

We’d like to discuss technical details related to a specific cloud in a SIG meeting with people who are also involved into development in this domain, and that way gain useful insights into the cloud infrastructure, improve the overall quality of our features, stay on top of the new features, and improve the relations with maintainers outside of our company, which nevertheless share with us a common goal.

### Implementation Details

> The `external` `cloud-controller-manager` implementation will be hosted within openshift repository in https://github.com/openshift/cluster-cloud-controller-manager-operator. That repository will manage `CCM` provisioning for metioned cloud providers + `Openstack` out-of-tree provider implementation described in the [document](https://github.com/openshift/enhancements/tree/master/enhancements/cloud-controller-manager/openstack-cloud-controller-manager.md)

Reviewer 6: > The `external` `cloud-controller-manager` implementation 

I cannot follow. This is the operator or the operand? As ccm implementation I would be the latter. But called operator?

Reviewer 3: I think this is referring to an operator to deploy the CCM, if we take that route.

Probably not worth calling out openstack separately, cloud providers in this sense encompasses any platform I think

> `Cloud-controller` will be provisioned in the `openshift-cloud-controller-manager` namespace.

Reviewer 3: What is `Cloud-controller`? Is this referring to the `cloud-controller-manager` or the `cluster-cloud-controller-manager` operator that has been discussed?

Author: Will decide on namespaces in https://github.com/openshift/enhancements/pull/463/files#r498377175

Preparation steps, common for all providers:

> - Add support for recognition of the `external` option in [library-go](https://github.com/openshift/library-go/blob/master/pkg/operator/configobserver/cloudprovider/observe_cloudprovider.go) which will disable config parsing and add support for `external` field in the `Infrastructure` resource.

Reviewer 2: can we specify which existing components would be consuming this i.e kube-apiserver and kube-controller manager.

Reviewer 3: Note for kube-controller-manager and kube-apiserver, there is no external option, we just need to not specify the flag (see [docs](https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/#running-cloud-controller-manager)).

This should be called out explicitly somewhere in this proposal IMO, quoting those docs ideally

> - Port the templates of `cloud-controller` from upstream repositories into proposed https://github.com/openshift/cluster-cloud-controller-manager-operator repository. Those templates will belong to the newly created `openshift-cloud-controller-manager` namespace and be managed with an operator.

Reviewer 2: Which templates is this referring to,  can we please elaborate this a bit?

Reviewer 3: Agreed, would be good to link here to the various templates for the providers so that we have an understanding of what we are actually going to be deploying here

> - Operator will be deployed in the `openshift-cloud-controller-manager-operator` namespace.

Reviewer 3: Why do we need to deploy this to a separate namespace, I would expect the operator and operands to be in the same namespace as with other components within openshift (eg MAO)

Author: https://github.com/openshift/enhancements/pull/463/files#r497488799

> Each provider implementation will need to do the following minimal set of actions in order to make out-of-tree implemetation work:

Reviewer 3: >IIUC we'll also need to import the CCM repositories for each supported platform under the openshift namespace and adapt the manifests to make them work in openshift.

@Danil-Grigorev If this isn't already noted in the doc somewhere, can we add it in? If it is already we can resolve this comment

> - Use `external` flag in the `kube-controller-manager` pod instead of the previously used provider. Remove the `cloud-config` option from the template.

Reviewer 3: There is no external flag, this should say `remove the --cloud-provider` argument

Author: The official docs are incorrect :D

> - Use `external` flag in the `kubelet` pod. Remove the `cloud-config` option from the template.

Reviewer 3: We can be more specific by what we mean here
```suggestion
- Use `--cloud-provider=external` flag in the `kubelet` pod. Remove the `cloud-config` option from the template.
```

> - Provision a `cloud-controller-manager` `DaemonSet` in the `openshift-cloud-controller-manager` namespace.

Reviewer 2: Why a daemonSet? I suggest a deployment unless there's a reason for us to use static pod, i.e if we need it for the cluster to bootstrap or come back up.

Reviewer 6: Not even a DaemonSet is about static pods.

My questions:
- can we disaster-recover a cluster without ccm running?
- what is non-functional in a cluster during bootstrapping until ccm is up?
- who talks to ccm? Only kcm or also ks and ka?
- what are the version skew compatibility constraints? Can n+1 kcm talk to n ccm and conversely?
- does ccm provide an API? How is it hosted? https endpoint? Through load balancer?
- what are the thoughts about certificate management?

Reviewer 3: > can we disaster-recover a cluster without ccm running?

Without CCM, any new kubelet remains tainted as uninitialised, from the docs:
>> kubelets specifying --cloud-provider=external will add a taint node.cloudprovider.kubernetes.io/uninitialized with an effect NoSchedule during initialization. This marks the node as needing a second initialization from an external controller before it can be scheduled work. Note that in the event that cloud controller manager is not available, new nodes in the cluster will be left unschedulable. The taint is important since the scheduler may require cloud specific information about nodes such as their region or type (high cpu, gpu, high memory, spot instance, etc).

So this would affect recovery if new nodes are needed to be booted

> what is non-functional in a cluster during bootstrapping until ccm is up?

Based on the above, pods will not be able to schedule unless they tolerate the uninitialised taint, while we could do this, without initialised nodes, we don't have any availability information and as such affinities based scheduling decisions are not possible

The above makes me feel like it's a pretty fundamental controller and lean towards running the CCM as a static pod in a similar way to KCM, I'm not entirely sure of the implications of doing this though so would be keen to know if there have previously been discussions about what warrants running as a static pod

Reviewer 6: @Danil-Grigorev can you answer the other questions in the list? They are fundamental to asses possible architectures.

Reviewer 6: More questions:
- what happens if the kcm leader has no access to ccm? Can it continue its work? Will it give up leadership?
- how does a kubelet talk to the ccm? Does every node need a ccm? What about worker nodes?

Author: Hey! I tried to answer all the questions in the FAQ section of the proposal. Please unresolve it if you think it is necessary

#### API resource changes

> In order to achieve desired cluster configuration with `external` cloud components, we need to add a new field into `Infrastructure` [resource](https://github.com/openshift/api/blob/c2f7aea5d89ee87d7510b50bdb08dde9028a53ef/config/v1/types_infrastructure.go#L11). With the addition, a `status.platformStatus.external` field will determine the architectural change in the cluster configuration. Other components will be able to consume the configuration same way as before, only with a minor change of the resource processing.

Reviewer 2: who set this value and when?

Author: It is populated by the installer for every platform: https://github.com/openshift/installer/blob/0f625f6cc1028fd0a947dbee6c53b200c47466bf/pkg/asset/manifests/infrastructure.go#L56

Reviewer 6: infrastructure.status.platformStatus.external is a very generic name.

Reviewer 3: Why do we need an API field for this? Surely this is internal to the state of the cluster and we should be able to handle this without exposing it as an API field to the customer?

I had imagined everything would be internal or external based on the `platform` given a particular version of OCP, there wouldn't be a choice or logic within any controller to determine internal or external.

Where KCM and KA are configured, we need to return empty string and have the `--cloud-provider` flag ommitted, there's a switch that builds this already which could just be updated, I imagine as we upgrade to OCP n+1 there is a component that can reconfigure these?

Where `Kubelet` is configured, we need to set `--cloud-provider=external` based on a list of providers being external or not, I imagine there's some switch on `platform` already that determines the value for this flag, we just need to update that. Will MCO be able to update that on each node during an upgrade?

The operator or whatever deploys CCM will be able to look at the `platform`, and deploy the correct binary dependent on that.

> In order to achieve desired cluster configuration with `external` cloud components, we need to add a new field into `Infrastructure` [resource](https://github.com/openshift/api/blob/c2f7aea5d89ee87d7510b50bdb08dde9028a53ef/config/v1/types_infrastructure.go#L11). With the addition, a `status.platformStatus.external` field will determine the architectural change in the cluster configuration. Other components will be able to consume the configuration same way as before, only with a minor change of the resource processing.

Reviewer 2: How are we planning to maintain the lifecycle of the cloud controller manager and its config flags?
I'd expect this to come in hand with a new CRD similar to https://github.com/openshift/cluster-kube-controller-manager-operator/blob/master/bindata/v4.1.0/config/defaultconfig.yaml
https://github.com/openshift/api/blob/master/operator/v1/types_kubecontrollermanager.go

Reviewer 6: Don't create a config. Use flags only. The configs have been a sin of the past of OpenShift 3.x that we move away from. Just construct the flags and put them into the Deployment pod spec, compare openshift-apiserver, oauth-apiserver.

```go
type PlatformStatus struct {
  ...
	// External is a value used in infrastucture automaton to determine component-specific
	// behavior for the PlatformType value. Components depending on this variable will
	// ignore cloud specific detailes associated with the given platform type,
	// and will concider the value of the infrastructure to be "external" - essentially
	// disabling any further evaluation.
	//
	// This will allow to delegate task of processing platform specific bits
	// to another component.
	// +optional

>   External bool `json:"external"`

Reviewer 6: ExternalCloudControllerManager.

Why should that be in a status?

Reviewer 3: Is this field just being used as a temporary trigger during the development cycle? Will it ever make it into a production cluster?

If we need some way to signal that something is external while development is happening, an annotation would be preferred. Though I would expect we wouldn't need this signal in a production cluster once each of the constituent parts is aware that a particular platform is now considered external

  ...
```

> To ensure a smooth transition from the `in-tree` selection, `library-go` [implementation](https://github.com/openshift/library-go/blob/master/pkg/operator/configobserver/cloudprovider/observe_cloudprovider.go) will include an additional argument for the `cloudProviderObserver` structure, which will allow to preserve initial configuration strategy, until the new `observeExternal` flag will be set to `false`:

Reviewer 6: who will set observeExternal and when?

Author: There are a couple of PRs I opened to test the approach. This option will be set on `false` for all components currently watching the `cloud-config` file and relying on `library-go` implementation once they will be merged:
- https://github.com/openshift/library-go/pull/895
- https://github.com/openshift/cluster-kube-apiserver-operator/pull/953
- https://github.com/openshift/cluster-kube-controller-manager-operator/pull/450

These I assume should be merged into 4.7 branch to introduce the transitional phase.

Reviewer 3: I think this paragraph needs an update now we are leaning towards this temporary annotation approach

Author: It still applies, as it would be good to reuse the logic from the library-go implementation in both `kcm` and `ccm` operators. It will only rely on the annotation in this case.

```go

> type cloudProviderObserver struct {

Reviewer 2: I'm not sure why `observeExternal ` this is needed, we got all we need by watching at `status.platformStatus.external` to decide what to set for `cloudProviderConfigPath`

Reviewer 2: Is this the mechanism proposed to enable backward compatible changes during the development transition?

Author: Yes, this will be the first step to transition to our operator management. We will use this, so we won't diverge from the existing approach of watching and copying the `cloud-config` configMap into the namespace, where it should be managed. This is currently used by kube-api-server and kube-controller-manager operators. Later this will be substituted by a CR as you suggested, similar to the `KubeControllerManager` one.

Author: So it was decided this implementation will be for development/testing only. The final migration will be done by a set of changes on all the repos, reliant on the selected provider.

	targetNamespaceName     string
	cloudProviderNamePath   []string
	cloudProviderConfigPath []string
	observeExternal         bool
}
```

The `observeExternal` field will determine, if `status.platformStatus.external` field from the `Infrastructure` resource will be taken in account in establishing needed flags for a Kubernetes binary, such as some `cloud-controller-manager`, `kube-controller-manager`, etc. It will make the observer type to ignore content of the `cloud-config` file, and force the provider to be set as `external` if the option is not selected.

> #### Kubelet

Reviewer 8: Do the various component need to be updated in a specific order, eg `kcm` then `kubelet`, then `kube-apiserver`?  Are we expecting them to converge to a stable state if there is no order?

Author: The order for updating the existing component should not change much, we only need to ensure the `CCM` is running before or shortly after the `kubelet`. It would be great to deploy and run the operator at the same time as `kube-controller-manager` operator is starting.

Openshift manages `kubelet` configuration with `machine-config-operator`. The actual `cloud-config` arguments are passed externally from the `Infrastructure` resource on https://github.com/openshift/machine-config-operator/blob/2a51e49b0835bbd3ac60baa685299e86777b064f/pkg/controller/template/render.go#L310-L357, where the support for `external` cloud provider will be added. For each platform with `external` configuration, the `cloud-config` flag will not be set.

#### Kube-controller-manager

This configuration is managed by the `cluster-kube-controller-manager-operator` code and will need to integrate the `library-go` changes, described above in the operator [configuration](https://github.com/openshift/cluster-kube-controller-manager-operator/blob/990ea3f2ace13be6579c00a889842c3f5d3e756a/pkg/operator/configobservation/configobservercontroller/observe_config_controller.go#L82-L85).

#### Kube-api-server

> No changes for the current configuration are needed

Reviewer 2: doesn't the kube api server also need to point to `external`?

Author: You are absolutely right, just for some reason in the clusters I used those flags were not set so assumed this. There is actually a PR for this already, will add the detailed description in a minute - https://github.com/openshift/cluster-kube-apiserver-operator/pull/953

### Operator resource management strategy

Every cloud manages it's configuration differently, but share a common overall structure.

1. `cloud-controller-manager` and related resources are deployed under one namespace.

> 2. Each pod (`ccm`, `cloud-node-manager`, etc.) supports deployment with a `DaemonSet`.

Reviewer 2: why is this a daemonSet? https://github.com/openshift/enhancements/pull/463/files#r490110395

Reviewer 6: We have more than just ccm? Now I read cnm.

Author: It is indeed the case. Azure decided to split their `ccm` and `cnm` into separate binaries: https://github.com/kubernetes-sigs/cloud-provider-azure/tree/master/cmd

3. The pods are expected to run on `Nodes` marked as a control-plane with annotation:
```yaml
    node-role.kubernetes.io/master: ""
```

> 4. The pods are sharing common command line arguments with the `kcm` pods, such as `--cloud-provider` and `--cloud-config`.

Reviewer 3: Are we not removing this from KCM?

Author: We are, once it is working for each provider. Just out-of-scope for this enhancement.

> Proposed operator will not diverge from the upstream resource structuring:

Reviewer 6: What does this mean?

Author: Sorry, I only meant to say the operator will deploy the same set of workloads decided to be used in the upstream repositories. This way our fork for \<some\>cloud-provider will not diverge in the structuring from the upstream until it is decided otherwise.

If they are deploying additional service for some newly introduced feature or set a new port for metrics, we'd like to follow their implementation and avoid possible collisions with our own components, previously structured differently

Reviewer 6: This sounds ambitious. Expect that our manifests we seriously diverge from anything upstream provides. We might follow loosely their structure, but won't be able to use their manifests directly.

> - Static infrastructure resources creation, such as namespaces or service accounts will be delegated to the [CVO](https://github.com/openshift/cluster-version-operator).

Reviewer 2: https://github.com/openshift/enhancements/pull/463/files#r490122199

Author: Seems that a `Deployment` is a better choice :+1:

> - Operator will populate, create and manage the cloud-specific set of `DaemonSets`, based on the selected `platform` in the `Infrastructure` resource.

Reviewer 6: There are cloud specific workloads? This is a strong argument against merging it into kcm pod.

Author: Yes there are. Aside from `ccm` and `cnm` for Azure, there is also a VSphere with an additional node-port service for connecting to the API endpoint: https://github.com/kubernetes/cloud-provider-vsphere/blob/master/manifests/controller-manager/vsphere-cloud-controller-manager-ds.yaml#L63

Reviewer 3: I agree that this shouldn't be merged into the KCM pod, we should have our own, and have it cloud specific (eg called `aws-cloud-controller-manager` on AWS) so that if we ever decided to run more than one cloud in a cluster, the migration to that approach would not cause conflicts with naming

- The `--cloud-provider` and the `--cloud-config` arguments will be populated using existing `library-go` based implementation from the [cluster-kube-controller-manager-operator](https://github.com/openshift/cluster-kube-controller-manager-operator/blob/990ea3f2ace13be6579c00a889842c3f5d3e756a/pkg/operator/configobservation/configobservercontroller/observe_config_controller.go#L82-L85).

> 

Reviewer 3: If this is only going to be used by the CCCM or operator for the CCM pods whatever that is, does it really need to be shared in library-go?

Author: Once we get all providers `external`, we should consider refactoring or removing this.

#### Examples for the cloud-controller-manager configurations

##### AWS

Main repository: https://github.com/kubernetes/cloud-provider-aws/
Sample manifests: https://github.com/kubernetes/cloud-provider-aws/tree/master/manifests

> To try external cloud-controller-manager, simply run:

Reviewer 3: This mentions trying this, is that a phrasing from upstream or do they expect that these resources are production ready and this is how they'd recommend deploying in production?

Author: As Stefan mentioned, we have a completely different set of goals for Openshift, so their decisions are not applying to us. I removed this section for that reason, and we should probably settle for the combination of static pods during bootstrap and a Deployment at runtime phase.

```bash
oc apply -f https://raw.githubusercontent.com/kubernetes/cloud-provider-aws/master/manifests/rbac.yaml  
oc apply -f https://raw.githubusercontent.com/kubernetes/cloud-provider-aws/master/manifests/aws-cloud-controller-manager-daemonset.yaml
```

##### Azure

Main repository: https://github.com/kubernetes-sigs/cloud-provider-azure
Sample manifests: https://github.com/kubernetes-sigs/cloud-provider-azure/tree/master/examples/out-of-tree

##### GCP

Main repository: https://github.com/kubernetes/cloud-provider-gcp
Sample manifests: https://github.com/kubernetes/cloud-provider-gcp/tree/master/deploy

##### vSphere

Main repository: https://github.com/kubernetes/cloud-provider-vsphere
Sample manifests: https://github.com/kubernetes/cloud-provider-vsphere/tree/master/manifests/controller-manager

> ## Upgrade/Downgrade strategy

Reviewer 2: can we elaborate a downgrade scenario? What happens if we roll back after the new operator is running?
I'd expect the config to be restored for all components and the new operator to no-op.

Reviewer 3: I think this is another argument for not making this a field somewhere and just having the logic within the controllers based on the `platform`

> While a set of `cloud-controller-manager` pods is running under a `DaemonSet` provisioned by the operator, aquiring leadership over cloud specific control loops in a HA openshift deployment will be managed by a cross-namespace leader-election migration.

Reviewer 2: why is this a daemonSet? https://github.com/openshift/enhancements/pull/463/files#r490110395

> While a set of `cloud-controller-manager` pods is running under a `DaemonSet` provisioned by the operator, aquiring leadership over cloud specific control loops in a HA openshift deployment will be managed by a cross-namespace leader-election migration.

Reviewer 6: what is cross-namespace leader election?

Author: A fancy name, which came to mind while looking into leader-migration proposal - https://github.com/kubernetes/enhancements/blob/473d6a094f07e399079d1ce4698d6e52ddcc1567/keps/sig-cloud-provider/20190422-cloud-controller-manager-migration.md#proposal

Upstream [proposal](https://github.com/kubernetes/enhancements/blob/473d6a094f07e399079d1ce4698d6e52ddcc1567/keps/sig-cloud-provider/20190422-cloud-controller-manager-migration.md#motivation) describes the main implementaion details for this component. Their proposed implementation for cross-namespace leadership delegation would help Openshift to preserve HA during upgrage from in-tree to the out-of-tree.

In case the cluster does not require to be HA, without integrating this feature an upgrade under bare CVO management will lead to two simultanious replicas of `cloud-controllers` running at the same time at some moment.

### Test Plan

- Each provider at the time of migration should have a working CI implemetation, which will assist in testing the provisioning with out-of-tree support enabled.

> - Ensure that transition between the in-tree and the out-of-tree will be handled by the `Infrastracture` `status.platformStatus.external` field being set.

Reviewer 3: I think this transition should be based on going through an upgrade and not going through some option

> - Ensure the in-tree and the out-of-tree providers could co-exist in the cluster if this is supported.

Reviewer 3: We should be aiming to not support this for a single provider if that makes sense, we should cut over on a 4.n to 4.n+1 boundary

> - Ensure that upgrade from a release running on the in-tree provider only to the out-of-tree succeeds.

Reviewer 2: Can we elaborate the CI jobs we are planning to run over the new operator repo and the ones for providers?

## Other notes

##### Removing a deprecated feature

> Upstream is currently planning to remove support for the in-tree providers with OCP-4.8 (kubernetes 1.21) - based on the slack [conversation](https://kubernetes.slack.com/archives/CPPU7NY8Y/p1589289501020600). The enchancement [describes](https://github.com/kubernetes/enhancements/blob/473d6a094f07e399079d1ce4698d6e52ddcc1567/keps/sig-cloud-provider/20190125-removing-in-tree-providers.md) steps towards this process. Our goal is to have a working alternative for the in-tree providers ready to be used on-par with old implementation befor the upstream release will remove this feature from core Kubernetes repository.

Reviewer 6: does this mean we must be finished with the implementation of this enhancement in 4.7 order to allow:
- 4.7 -> 4.6 downgrade: 4.6 won't have ccm support. So in 4.7 we have to keep using internal ccm.
- 4.8 -> 4.7 downgrade: 4.7 knows both internal and external. So we can switch to external only in 4.8 and allow a downgrade to 4.7 with external ccm.

Author: Yes, this sounds right. Openstack plans to ship their optional support in 4.7, so this will be supported, but won't be default at this point. https://github.com/openshift/enhancements/pull/423/files#diff-f511056b9f0aeb32f492f7c41c08cadbR132 But I would refer to @Fedosin for details on that.

Reviewer 3: Why are we making this an optional feature? Does it need to be optional?

Author: Added a more detailed section about upgrades/downgrades for each release. Sorry for the confusion, for OpenStack it will be default in 4.7.

> ## Timeline

Reviewer 6: this is expecte to have a timeline for our implementation, not for upstream. Please add, especially answering questions around https://github.com/openshift/enhancements/pull/463/files#r492579924.

Follow the Kubernetes community discussions and implementation/releases

### Kubernetes 1.18

- vSphere support graduaded to beta: https://github.com/kubernetes/enhancements/issues/670

### Kubernetes 1.19

- Azure support goes into beta: https://github.com/kubernetes/enhancements/issues/667

## Infrastructure Needed

### New github projects:

Forks for out-of-tree provider repositories:
- [AWS](https://github.com/kubernetes/cloud-provider-aws)
- [GCP](https://github.com/kubernetes/cloud-provider-gcp)
- [Azure](https://github.com/kubernetes-sigs/cloud-provider-azure)

> - [Vsphere](https://github.com/kubernetes-sigs/cloud-provider-vsphere)

Reviewer 2: broken link

- Define and implement a downgrading path from newer OpenShift versions with the CCM to those that use the in-tree cloud provider.

### Non-Goals

- [CSI driver migration](https://kubernetes.io/blog/2019/12/09/kubernetes-1-17-feature-csi-migration-beta/) is out of scope of this work.

## Proposal

Our main goal is to start using Cloud Controller Manager in OpenShift 4, and make a seamless transition for all currently supported providers that require `cloud provider` interface: OpenStack, AWS, Azure, GCP and vSphere.

To maintain the lifecycle of the CCM we want to implement a cluster operator called `cluster-cloud-controller-manager-operator`, that will handle all administrative tasks: deploy, restore, upgrade, and so on.

### Implementation Details

The `cluster-cloud-controller-manager-operator` implementation will be hosted within OpenShift repository in https://github.com/openshift/cluster-cloud-controller-manager-operator. That repository will manage CCM provisioning for mentioned cloud platforms.

- All platform CCMs will be provisioned in `openshift-cloud-controller-manager` namespace.

> - The Operator will be deployed in `openshift-cloud-controller-manager-operator` namespace.

Author: Probably one will be enough, but it depends on which operator we will pick as an example one. So far the [kube-controller-manager-operator](https://github.com/openshift/cluster-kube-controller-manager-operator) seems the fittest, and it spawns it's management across two namespaces - `openshift-kube-controller-manager` and `openshift-kube-controller-manager-operator`.

2. Each pod (`CCM`, `cloud-node-manager`, etc.) will be managed by a `Deployment`.
3. The pods are expected to run on the control-plane `Nodes` using `hostNetwork`.

Operator will manage:

- Static infrastructure resources creation, such as namespaces or service accounts will be delegated to the [CVO](https://github.com/openshift/cluster-version-operator).
- Operator will populate, create and manage the cloud-specific set of `Deployments`, based on the selected `platform` in the `Infrastructure` resource.
- The `--cloud-provider` and the `--cloud-config` arguments will be populated using existing `library-go` based implementation from the [cluster-kube-controller-manager-operator](https://github.com/openshift/cluster-kube-controller-manager-operator/blob/990ea3f2ace13be6579c00a889842c3f5d3e756a/pkg/operator/configobservation/configobservercontroller/observe_config_controller.go#L82-L85) to share common approach with the bootstrap specific static-pod configuration.
- Operator will be build with a goal to achieve provider decoupling and potential switching between providers in the cluster.

## Upgrade/Downgrade strategy

> A set of `cloud-controller-manager` pods will be running under a `Deployment` resource, provisioned by the operator. The leader-election will preserve the leadership during the cluster updates over cloud-specific loops.

Author: Yes, will revert the change with the removed section about lease migration: https://github.com/kubernetes/enhancements/blob/master/keps/sig-cloud-provider/20190422-cloud-controller-manager-migration.md

Upgrade from previous versions of OpenShift on OpenStack will look like:

- On the initial stage of upgrading `cluster-version-operator` starts `cluster-cloud-controller-manager-operator`.

> - `cluster-cloud-controller-manager-operator` creates a static pod manifest for CCM.

Author: I'll extend the bootstrap section with an explanation. Shortly, the reasoning is also explained in the Deployment `vs` static pod comparison. We need CCMs to be a part of the bootstrap procedure, so they should run before any of the master Nodes become ready to schedule pods, or the etcd being healthy.

- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

## Open questions

1. Should the post-install phase of the CCMs be managed by a static-pod, similarly as it is done for the bootstrapping phase, or with a `Deployment`?

Pros and cons:

Deployment:
+ Simple lifecycle management
+ Simple configuration changes
- Requires additional RBAC configurations and running kubernetes components
- Requires integration with the cloud credentials operator to access cloud credentials

Static pod:
+ Does not require a running/healthy etcd, api server and kube-controller-manager
+ Thus will be used during bootstrapping phase
- Requires management with an operator
- Limited configuration options, and could not be reused from the bootstrapping template, each cloud will require its own template

2. CCM relation to other openshift components, such as SDN and storage? How a non-operational CCM will affect cluster health, which components will take a hit?

- Currently known `kubelet` dependency added requirement to run the component at bootstrapping phase.
- Storage CSI migration will be required for OpenStack and vSphere providers
- Networking:

>   - In GCP and OpenStack the `Routes` [interface](https://github.com/kubernetes/cloud-provider/blob/0dc0419ec04907f9deb6aa498cece84f10044726/cloud.go#L63) is implemented. Does it have any relation to the SDN? [Link](https://github.com/kubernetes/cloud-provider/blob/master/controllers/route/route_controller.go) to the controller managing this.

Reviewer 4: AFAIK, the CloudProvider Routes() API is only used in GCE, when using the default GCE network plugin. Maybe there might be other third-party network plugins that use it in some cases? However, I don't think we provide any way to enable it in OCP; it only gets used if you pass `--configure-cloud-routes` to kube-controller-manager, and our [default config](https://github.com/openshift/cluster-kube-controller-manager-operator/blob/master/bindata/v4.1.0/config/defaultconfig.yaml) sets it false and I don't know if there's any way to override that.

Anyway, openshift-sdn and ovn-kubernetes definitely don't care about the Routes API at all. For clouds/plugin combinations where it does matter, it affects creation/deletion of nodes, so if CCM was unavailable it would mean that newly-created nodes would not become Ready (until CCM started / came back).

Author: `Routes()` could be enabled for use in [Azure](https://github.com/kubernetes/kubernetes/blob/c0071c617325642bd24d9b0587273bf7a7c7db42/staging/src/k8s.io/legacy-cloud-providers/azure/azure.go#L685) and [AWS](https://github.com/kubernetes/kubernetes/blob/82bdba99070b5ab066b42ece420d4c1c8b4bf439/staging/src/k8s.io/legacy-cloud-providers/aws/aws.go#L1418) as well, but it is disabled for all of them, as you said.

>   - OpenStack Load Balancing is an optional feature, though is [supported](https://github.com/kubernetes/cloud-provider/blob/master/controllers/service/controller.go) by the CCM interface with the [LoadBalancer()](https://github.com/kubernetes/cloud-provider/blob/0dc0419ec04907f9deb6aa498cece84f10044726/cloud.go#L49]. How important is this functionality during cluster runtime?

Author: Asked this question in https://coreos.slack.com/archives/CDCP2LA9L/p1601548483158900

>   - OpenStack Load Balancing is an optional feature, though is [supported](https://github.com/kubernetes/cloud-provider/blob/master/controllers/service/controller.go) by the CCM interface with the [LoadBalancer()](https://github.com/kubernetes/cloud-provider/blob/0dc0419ec04907f9deb6aa498cece84f10044726/cloud.go#L49]. How important is this functionality during cluster runtime?

Reviewer 4: So (as I understand it) we require that either the cluster has a CloudProvider that implements LoadBalancer services (which is how most OCP clusters work), or else that you provide some other way to make (OCP) Routes work (which is what we do on bare metal, and I guess some OpenStack installs?). If the cluster is relying on CloudProvider LoadBalancer support, then if the CCM went away, then changes to LoadBalancer Services would not be processed until it came back.

I don't think we try to create any LoadBalancer Services at install time until after the worker nodes are created (at which point cluster-ingress-operator will be trying to create the default router), so as long as CCM is working by then, things should work. I'm not sure if ingress/routes creates new any LoadBalancer Services _after_ install time. If they did, and CCM was unavailable, then the ingress/route would not be usable until CCM came back to set up the load balancer. Of course if the user is creating LoadBalancer Services of their own, then that would also fail if CCM wasn't available.

Maybe need some Network Edge input here

Author: From the slack [discussion](https://coreos.slack.com/archives/CCH60A77E/p1601565268029300) @Miciah confirmed this. The `router-default` LB is created for the ingress-operator deployment, so having CCM ready at this point and reporting available condition + no alerts firing guaranties that everything will be working.

- `kube-apiserver` and `kube-controller-manager` are restarted with the `--cloud-provider=some-platform` option.

- `cluster-cloud-controller-manager-operator` stops working.

### Version Skew Strategy

See the upgrade/downgrade strategy.

### Action plan (for OpenStack)

#### Build OpenStack CCM image by OpenShift automation

To start using OpenStack CCM in OpenShift we need to build its image and make sure it is a part of the OpenShift release image. The CCM image should be automatically tested before it becomes available.
The upstream repo provides the Dockerfile, so we can reuse it to complete the task. Upstream image is already available in [Dockerhub](https://hub.docker.com/r/k8scloudprovider/openstack-cloud-controller-manager).

Actions:

- Configure CI operator to build OpenStack CCM image.

CI operator will run containerized and End-to-End tests and also push the resulting image in the OpenShift Quay account.

#### Test Cloud Controller Manager manually

When all required components are built, we can manually deploy OpenStack CCM and test how it works.

Actions:

> - Manually install CCM's [daemonset](https://github.com/kubernetes/cloud-provider-openstack/blob/master/manifests/controller-manager/openstack-cloud-controller-manager-ds.yaml) on a working OpenShift cluster deployed on OpenStack.

Reviewer 3: Should we update this to a deployment to more closely mimic what the operator would do?

Author: They use a `daemonSet` for that example, and I assume will yet require some changes for the config maps mounts used in that pod configuration, as well as other arguments. Do you think this section should be included in the proposal, @Fedosin?

## Motivation

The Kubernetes community is currently moving towards exclusion of the in-tree cloud-provider code from the core Kubernetes repository, and instead introducing out-of-tree provider support, making them independent from the core Kubernetes release cycle, and simplifying the process of development and e2e testing for those plugins. Previously each cloud-provider was located in the [`legacy-cloud-providers`](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/legacy-cloud-providers) and now the support and main development is done in the dedicated cloud-provider [repositories](https://github.com/kubernetes?q=cloud-provider-). They plan to remove the in-tree support completely in the future releases, approximetly in [v1.21](https://kubernetes.slack.com/archives/CPPU7NY8Y/p1589289501020600).

### Goals

- Prepare openshift components to accomodate the out-of-tree plugins and remove support for the in-tree implementation.
- Describe an approach how an `external` cloud provider configuration for the cloud will be supported.

> - Provide the means to allow generic `CCM` component registration via the provider specific upstream implemetation.

Reviewer 3: Not sure what this means, perhaps we can elaborate on this or reword to make it clearer?

> - Concider support for running switching cloud configurations in a single cluster.

Reviewer 3: Is this as in, switching from GCP to AWS? If so, I don't think that should be a goal of this project

Or is this switching from in-tree to out-of-tree as an option? Also don't think that should be a goal here

Minor typo btw
```suggestion
- Consider support for running switching cloud configurations in a single cluster.
```

### Non-Goals

- Force immediate exclusion of in-tree support for currently used providers, as their out-of-tree counterparts are added.
- Add support for CSI and other out-of-tree implementations of cluster storage interfaces.

## Proposal

Add an out-of-tree support based on current cluster infrastructure, and make a seamless transition for all currently supported providers.

> This change will be gradually executed on all suported cloud providers: OpenStack, AWS, Azure, GCP and vSphere. The change could help inclusion for other cloud providers, such as [Yandex](https://github.com/flant/yandex-cloud-controller-manager) and [IBM](https://github.com/kubernetes/enhancements/issues/671) `CCMs` (`cloud-controller-managers`), if this will be concidered a goal.

Reviewer 3: Couple of typos
```suggestion
This change will be gradually executed on all supported cloud providers: OpenStack, AWS, Azure, GCP and vSphere. The change could help inclusion for other cloud providers, such as [Yandex](https://github.com/flant/yandex-cloud-controller-manager) and [IBM](https://github.com/kubernetes/enhancements/issues/671) `CCMs` (`cloud-controller-managers`), if this will be considered a goal.
```

### User Stories

#### Story 1:

> As a cloud developer, I’d like to improve support for cloud features for openshift. I'd like to develop and test new cloud providers with do not ship with default kubernetes distiribution, and require support for external cloud controller manager.

Reviewer 3: Think you mean which here
```suggestion
As a cloud developer, I’d like to improve support for cloud features for openshift. I'd like to develop and test new cloud providers which do not ship with default kubernetes distiribution, and require support for external cloud controller manager.
```

#### Story 2:

> As a developer I'd like to develop, build and release fixes independently from the kubernetes core, and assume they will first land upstream and then will be carried over into openshift destribution with less effort.

Reviewer 3: Minor improvement
```suggestion
As a developer I'd like to develop, build and release fixes for cloud provider controllers independently from the kubernetes core, and assume they will first land upstream and then will be carried over into openshift distribution with less effort.
```

#### Story 3:

We’d like to discuss technical details related to a specific cloud in a SIG meeting with people who are also involved into development in this domain, and that way gain useful insights into the cloud infrastructure, improve the overall quality of our features, stay on top of the new features, and improve the relations with maintainers outside of our company, which nevertheless share with us a common goal.

### Implementation Details

> The `cloud-controller-manager-operator` implementation will be hosted within openshift repository in https://github.com/openshift/cluster-cloud-controller-manager-operator. That repository will manage `CCM` provisioning for metioned cloud providers.

Reviewer 3: Are we going with `cluster-cloud-controller-manager-operator` for the name? If so we should update this reference.

```suggestion
The `cluster-cloud-controller-manager-operator` implementation will be hosted within openshift repository in https://github.com/openshift/cluster-cloud-controller-manager-operator. The operator will manage `CCM` provisioning for supported cloud providers.
```

- `CCM` will be provisioned in the `openshift-cloud-controller-manager` namespace.

> - Operator will be deployed in the `openshift-cloud-controller-manager-operator` namespace.

Reviewer 3: What's the motivation behind separate namespaces?

Author: Just trying to follow the pattern from `kcm` operator. But I guess there is no need in it, let's use the `openshift-cloud-controller-manager` namespace.

Preparation steps, common for all providers:

- Add support for recognition of the `external` configuration for each provider in [library-go](https://github.com/openshift/library-go/blob/master/pkg/operator/configobserver/cloudprovider/observe_cloudprovider.go).
- Port the templates of `cloud-controller` from upstream repositories into proposed https://github.com/openshift/cluster-cloud-controller-manager-operator repository. Those templates will belong to the newly created `openshift-cloud-controller-manager` namespace and be managed with an operator.

Each provider implementation will need to do the following minimal set of actions in order to make out-of-tree implemetation work:

- Use `external` flag in the `kube-controller-manager` pod instead of the previously used provider. Remove the `cloud-config` option from the template.
- Use `external` flag in the `kubelet` pod. Remove the `cloud-config` option from the template.
- Provision a `cloud-controller-manager` `DaemonSet` in the `openshift-cloud-controller-manager` namespace.

#### Resource changes

##### Pre-release / Development

> In order to achieve desired cluster configuration with `external` cloud components, we need to add an annotation on the `Infrastructure` [resource](https://github.com/openshift/api/blob/c2f7aea5d89ee87d7510b50bdb08dde9028a53ef/config/v1/types_infrastructure.go#L11). The annotation will determine the architectural change in the cluster configuration. Other components will be able to consume the configuration same way as before, and will decide which flags will be set, if this annotation would be found.

Reviewer 3: Might make it clearer the value here?
```suggestion
In order to achieve desired cluster configuration with `external` cloud components, we need to add an annotation on the `Infrastructure` [resource](https://github.com/openshift/api/blob/c2f7aea5d89ee87d7510b50bdb08dde9028a53ef/config/v1/types_infrastructure.go#L11). The annotation will determine the architectural change in the cluster configuration. Other components will be able to consume the configuration same way as before, and will decide how to deploy cloud controllers based on the value of this annotation.
```

Proposed annotation: `infrastructure.openshift.io/external-cloud-provider: true`.

> The annoation will be a temporary measure to simlify the operator development, and enable testing the transition in CI. Once all providers will become out-of-tree, the annotation will be safe to remove.

Reviewer 3: Trying to expand on the value and what we gain by adding this annotation
```suggestion
The annotation will be a temporary measure to simplify the operator development allowing multiple components to merge implementations of this feature independently before switching on the feature by changing the value of the annotation. Once all providers are moved out-of-tree, the annotation will be safe to remove.
```

The change will help adding support for other cloud platforms, such as [Packet](https://github.com/packethost/packet-ccm), [Digital Ocean](https://github.com/digitalocean/digitalocean-cloud-controller-manager) or [Alibaba Cloud](https://github.com/kubernetes/cloud-provider-alibaba-cloud).

It's especially important to do this for OpenStack because switching to the external cloud provider fixes many issues and limitations with the in-tree cloud provider, such as it's reliance on [Nova metadata service](https://docs.openstack.org/nova/latest/admin/metadata-service.html). For OpenStack platform, this means the possibility for deploying on provider networks and at the edge.

### Goals

- Prepare OpenShift components to accommodate the CCMs instead of deprecated in-tree cloud providers.
- Provide the means to allow management of generic `CCM` component based on the platform specific upstream implementation.
- Define and implement an upgrading and downgrading paths between the in-tree and the out-of-tree cloud provider configurations.

> - Ensure the transition will ensure the feature parity between the in-tree and the out-of-tree provider.

Reviewer 6: What should this sentence mean? Why do they have to have feature parity?

> - Ensure the transition will ensure the feature parity between the in-tree and the out-of-tree provider.

Reviewer 3: I think this could be (based on my interpretation of this statement - we don't want to lose any functionality):
```
- Ensure full feature parity between the in-tree and the out-of-tree providers before migration to ensure core Kubernetes cloud features (eg Load Balancer Services) continue to operate as before.
```
But I wonder if that may be a tad long winded for a Goal, WDYT?

#### Kube Controller Manager and Kube API Server

Configuration of these components is managed by `cluster-kube-controller-manager-operator` and `cluster-kube-apiserver-operator` respectively. Both operators use [`configobserver`](https://github.com/openshift/library-go/blob/16d6830d0b80dc2a3315207116d009ed2dd4cebf/pkg/operator/configobserver/cloudprovider/observe_cloudprovider.go#L140) module from `library-go` where we set what cloud provider should be used for the given platform. When a platform is being switched to `CCM`, we [must not](https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/#running-cloud-controller-manager) set `cloudProvider` value there and leave it empty.

Next step is to bump `library-go` for the operators.

##### Pre-release / Development

In order to achieve desired cluster configuration for cloud components during development, we need to add an annotation on the `Infrastructure` [resource](https://github.com/openshift/api/blob/c2f7aea5d89ee87d7510b50bdb08dde9028a53ef/config/v1/types_infrastructure.go#L11).

> Proposed annotation: `infrastructure.openshift.io/external-cloud-provider`.

Reviewer 5: This is not going to be accepted by @openshift/api-approvers. We already sufficient information which platform we are running on. Based on that, we should gradually set appropriate cloud-config values for all components and enable the necessary elements of ccm. I imagine this will take several releases, but at any given release we will document which external providers are supported in-tree and which out-of-tree.

Reviewer 5: I doubt that there will be half-implemented solution, as proposed through annotation. For example, for openstack we'll move all components (kas, kcm, kubelet) to not to set cloud-config flag.

Reviewer 3: I think the intention of this is that it shouldn't make it into a release and that it shouldn't be publicly available or documented.

During the development cycle, we will need to update various components to work with out of tree providers, if we implement these one by one and switch (eg update KCM to remove the flag), then this will break clusters and will not pass E2E.

The idea is that once all of the components have implemented their changes, we will be able to test all of the components together by adding the annotation, somehow add it to new clusters as part of the install process, then update all of the components to remove the switch, and then remove the annotation. 

I believe this is what this section is trying to explain.

Do you know of a better way to co-ordinate the changes across multiple components like this?

Reviewer 3: Was discussing this with @Fedosin and @Danil-Grigorev again and now that the plans have changed slightly (OSP TP in 4.7) it looks like we are going to need some way for a user to opt-in to having an external cloud provider for 4.7. I imagine however we implement this, it can replace the usage of the annotation as we have described it in this proposal.

I'm not sure what the standard practice for this is within OpenShift, is there some resource a user can create or some place where they can opt in to tech previews such as this?

I think the process we want to have is:
- User indicates via some resource that they want to enable out of tree
- KCMO, KASO, MCO observe this change and reconfigure the respective components to enable out of tree
- CCCMO observes this change and deploys the relevant CCM Deployment

Since there will be no CSI migration enabled for the TP for OSP, this will have to come with a proviso that all existing PVs will be lost, not sure how we communicate that.

As Out of tree moves to GA in future releases, the above components will just look at the `platform` value to determine which configuration they should be using on a given platform and this resource/field will no longer be required.

I'm imagining some logic something like:

```
if (featureFlagIsSet && setOfPlatformsInDevOrTP.Has(infrastructure.Platform)) || setOfPlatformsInGA.Has(infrastructure.Platform) {
  // Deploy out of tree configuration for given provider
} else {
  // Deploy in tree configuration for provider
}
```

The question is, where does the feature flag live? How is that implemented?

Reviewer 6: o/api/config/v1 FeatureGate

The annotation will determine the architectural change in the cluster configuration. If this annotation would be found, other components, such as `kube-apiserver`, `kube-controller-manager` and `kubelet` will set their flags accordingly, to allow hosting external `CCM` for any platform.

The annotation will be a temporary measure to simplify the operator development allowing multiple components to merge implementations of this feature independently before switching on the feature by changing the value of the annotation. Once all providers are moved out-of-tree, the annotation will be safe to remove.

> To ensure a smooth transition from the `in-tree` selection, `library-go` [implementation](https://github.com/openshift/library-go/blob/master/pkg/operator/configobserver/cloudprovider/observe_cloudprovider.go) will include an additional argument for the `cloudProviderObserver` structure, which will allow to preserve initial configuration strategy, until the new `observeExternal` flag will be set to `false`:

Reviewer 5: Not relevant in proposal.

`CCM` operator will provide initial manifests that allow to deploy `CCM` on the bootstrap machine with `bootkube.sh` [script](https://github.com/openshift/installer/blob/master/data/data/bootstrap/files/usr/local/bin/bootkube.sh.template).

At the initial stage of CCM installation installer creates a config map with `cloud.conf` key that contains configuration of CCM in `ini` format. The contents of `cloud.conf` are static and generated by the installer:

Example for the OpenStack:
```txt
[Global]
secret-name = openstack-credentials
secret-namespace = kube-system
```

Resources deployed here will be destroyed with the bootstrap stage cleanup.

#### Operator resource management strategy

Every cloud manages it's configuration differently, but shares a common overall structure.

1. `cloud-controller-manager` and related resources are deployed under a single namespace.
2. Each pod (`CCM`, `cloud-node-manager`, etc.) will be managed by a `Deployment`.

> 3. The pods are expected to run on the control-plane `Nodes` using `hostNetwork`.

Author: Depoloyments provide capability to do rolling upgrades, thanks to ReplicaSets, and the DaemonSets does not. On the last call, we decided that it is sufficient to run a single pod, scheduled on a control-plane node, so we don't have to nessesary span one over each node.

> 3. The pods are expected to run on the control-plane `Nodes` using `hostNetwork`.

Reviewer 3: It was my understanding on the call that we had agreed to running a single Deployment with multiple (2) pods. Using anti-affinity during scheduling to make sure that the pods do not end up on the same machine providing some fault tolerance should a machine fail

> 3. The pods are expected to run on the control-plane `Nodes` using `hostNetwork`.

Author: Added a section about this in the enhancement, thanks @JoelSpeed :)

> 3. The pods are expected to run on the control-plane `Nodes` using `hostNetwork`.

Reviewer 6: DaemonSets are not evicted normally from nodes on reboot. If I remember right, we saw issues with kcm some releases back with not giving up its lease? @soltysh might remember and can give more info here.

- Operator will be built with a goal to achieve provider decoupling and potential switching between providers in the cluster, and will consider support to run multiple providers in a cluster simultaneously (long-term goal).

#### CVO managemnet

The operator image will be build and included in every release payload, and expected to be deployed and running at kuberentes operators level ([10-29](https://github.com/openshift/cluster-version-operator/blob/34010292c3abd2582b687e9c0ef76c5924998f39/docs/dev/operators.md#how-do-i-get-added-as-a-special-run-level)) right after the `kube-controller-manager-operator` at runlevel [25](https://github.com/openshift/cluster-kube-controller-manager-operator/tree/master/manifests).

Proposed `CCMO` manifests runlevel is `26`.

#### Cloud provider management

Each cloud provider will be hosted under its own repository. Those repositories will be initially bootstrapped by forking upstream ones. 

> The `CVO` will be responsible for provisioning static resources, such as `SA`, `RBAC`, `ServiceMonitors` for metrics for the operator. Then, the operator will  be responsible for constructing and creating cloud provider resources.

Reviewer 3: Should spell out `SA` as `ServiceAccounts` and `RBAC` as `Roles/RoleBindings`. No real need to abbreviate here IMO

Forked provider repositories will also contain cloud-provider specific code, and a set of binaries for each of them to build and run. Each provider will be build inside a single image, which will be included in the release payload. The responsibility for the operator will be to choose which provider image should be used in a cluster at any moment.

#### CCM migration from KCM pod

> Currently the cloud-specific controller loops are running inside the `kcm` `Pod`. We need to preserve leadership over these parts during migration process. This includes migration from the `openshift-kube-controller-manager` namespace to the `openshift-cloud-controller-manager` namespace, which is not possible with current `leader-election` implementation.

Reviewer 3: From the attached proposal:
```
The migration mechanism outlined in this KEP should only be used for Kubernetes clusters that have very strict requirements on control plane availability. If a cluster can tolerate short intervals of downtime, it is recommended to update your cluster with in-tree cloud providers disabled, and then deploy the respective out-of-tree cloud-controller-manager.
```
Are we absolutely sure we can't tolerate a minimal down time? What's the trade-off here? How would we migrate if we did it with a small amount of down time? Has this been explored at all?

Perhaps the CCCMO could lookup the KCM pod and check the configuration before deploying the new deployment? 

Would be good to have an alternatives section detailing this so we know that we have explored this idea and discarded it (if that is the case)

Author: Added a paragraph about the workaround we could use.

#### Credentials management

The operator is expected to be integrated with the [cloud-credentials-operator](https://github.com/openshift/cloud-credential-operator), and issue the fine grained credentials for the cloud components.

*Bootstrap phase:*

- Initial configuration with static pods is expected to be using `ConfigMap` credentials, similarly to what is [done](https://github.com/openshift/installer/blob/ba6d7fe087eada5a4fc260064c85a484a8c45aaf/data/data/bootstrap/files/usr/local/bin/bootkube.sh.template#L162) in `kcm` operator.
- Once it is supported, the static pods will also use the `CredentialsRequest` to get cloud credentials.

*Post-install phase:*

- Using `CredentialsRequest` is the default option, every supported `CCM` within the `Deployment` will request its own set of credentials.

> - A set of `CredentialsRequest` resources will be hosted under each provider repository, and created by the `CVO`, similarly to the [machine-api](https://github.com/openshift/machine-api-operator/blob/6f629682b791a6f4992b78218bfc6e41a32abbe9/install/0000_30_machine-api-operator_00_credentials-request.yaml) approach.

Reviewer 3: Should these be centralised in the CCCMO repo?

A: CCM config is backward compatible with the in-tree cloud provider. It means we can reuse it.

Q: How to migrate PVs created by the in-tree cloud provider?
A: [CSIMigration](https://kubernetes.io/blog/2019/12/09/kubernetes-1-17-feature-csi-migration-beta/) looks like the best option, especially if it is GA in 1.19.

Q: Does CCM provide an API? How is it hosted? HTTPS endpoint? Through load balancer?
A: No. The CCM vendors part of the code serving cloud features from the [cloud-provider](https://github.com/kubernetes/cloud-provider). CCM provides a secure TCP connection on port [10258](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cloud-provider/ports.go#L22) and communicates similarly to other kubernets components.

Q: What happens if the kcm leader has no access to ccm? Can it continue its work? Will it give up leadership?
A: Expect KCM perform independently to CCM and do not care about its state after initial code migration. The CCM cluster operator will be responsible for going Unavailable or Degraded in a similar situation.

Q: Can we disaster-recover a cluster without ccm running?

> A: Yes we can, only the Node management and Route provisioning will be compromized. 

Reviewer 3: This question seems to be a duplicate of the first one on L#378

Q: What are the thoughts about certificate management?
A: Certificate management should be performed by the CCM operator, like it's done in the KCM operator.

Q: What happens if the KCM leader has no access to CCM? Can it continue its work? Will it give up leadership?

> A: Since KCM does not communicate with CCM it can continue to work if CCM is not available.

Reviewer 3: This question seems to be a duplicate of the question on L#402

Q: Does every node need a CCM?
A: No. Control plane nodes only. [Source](https://kubernetes.io/docs/concepts/overview/components/#cloud-controller-manager)

> Q: How does SDN depend on CCM?

Reviewer 3: This could probably be merged with the question on L#387

3. Should we reuse the existing cloud provider config or generate a new one?
CCM config is backward compatible with the in-tree cloud provider. It means we can reuse it.

4. How to migrate PVs created by the in-tree cloud provider?
[CSIMigration](https://kubernetes.io/blog/2019/12/09/kubernetes-1-17-feature-csi-migration-beta/) looks like the best option, especially if it is GA in 1.19.

## Summary

This enhancement proposal describes the migration of cloud platforms from the deprecated [in-tree cloud providers](https://kubernetes.io/docs/concepts/cluster-administration/cloud-providers/#openstack) to [Cloud Controller Manager](https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/using-openstack-cloud-controller-manager.md#get-started-with-external-openstack-cloud-controller-manager-in-kubernetes) services that implement `external cloud provider` [interface](https://github.com/kubernetes/cloud-provider).

As a pioneer platform, it is proposed to use OpenStack.

## Motivation

> Using Cloud Controller Managers (CCMs) is the Kubernetes' [preferred way](https://kubernetes.io/blog/2019/04/17/the-future-of-cloud-providers-in-kubernetes/) to interact with underlying cloud platforms as it provides more flexibility and freedom for developers. It replaces existing in-tree cloud providers, which have been deprecated and will be permanently removed approximately in Kubernetes [v1.21](https://kubernetes.slack.com/archives/CPPU7NY8Y/p1589289501020600). But they are still used in OpenShift and we must start a smooth migration towards CCMs.

Reviewer 4: > approximately in Kubernetes [v1.21](https://kubernetes.slack.com/archives/CPPU7NY8Y/p1589289501020600).

that statement is from May; the current target is 1.23

Reviewer 9: Hi Dan! Thank you for the information!
Can you provide any links to confirm that the removal has been delayed, so we can include them in the proposal?

Author: I think it is related to the removal of the field completely in kubelet in 1.23 - https://github.com/kubernetes/kubernetes/pull/90408#issuecomment-670057176 Let me have a better look.

Reviewer 4: sorry, missed this question earlier. I don't have a link to an official source. Andrew Sy Kim (who made the Slack comment you linked to) mentioned the 1.23 target [here](https://github.com/kubernetes/enhancements/pull/1665#issuecomment-703140901) and then I asked him on Slack and he confirmed that's the current target, and it's changed since the conversation you linked to.

Author: Thanks! I'm adding the updated info to the proposal

#### Story 2

As a developer I'd like to implement, build and release fixes for cloud provider controllers independently from the Kubernetes core, and assume they will first land upstream and then will be carried over into OpenShift distribution with less effort.

#### Story 3

As an OpenShift developer responsible for the cloud controllers, I want to share more code with upstream Kubernetes in order to ease feature and bug fix contributions both ways.

#### Story 4

We’d like to discuss technical details related to a specific cloud in a SIG meeting with people who are also involved into development in this domain, and that way gain useful insights into the cloud infrastructure, improve the overall quality of our features, stay on top of the new features, and improve the relations with maintainers outside of our company, which nevertheless share with us a common goal.

### Implementation Details

The `cluster-cloud-controller-manager-operator` implementation will be hosted within its own OpenShift [repository](https://github.com/openshift/cluster-cloud-controller-manager-operator). The operator will manage `CCM` provisioning for supported cloud providers.

> - The Operator and all platform cloud controllers will be provisioned in `openshift-cloud-controller-manager` namespace.

Reviewer 5: Move this sentence to where you're describing the ccm-o details.

Reviewer 6: This is also a strange factoring. Usually the operands (= cloud controllers) live in a different namespace than the operator.

Reviewer 3: This is my fault, it was originally two namespaces but I was asking why we need two, MAO all lives in one namespace and that seems to work, so I had asked what the reason for two namespaces would be.

If there's a convention for having operators and operands in separate namespaces we can do that here, is there a particular reason for it other than convention?

Author: We would also avoid questions about why an empty namespace exists in every cluster, while it does not host any operands in it. Having a single namespace seems to be clean an consistent to me too.

Reviewer 5: empty namespace doesn't cost anything

#### Resource changes

#### Kubelet

OpenShift manages `kubelet` configuration with `machine-config-operator`. The actual `cloud-config` arguments are set externally, based on the `Infrastructure` [resource values](https://github.com/openshift/machine-config-operator/blob/2a51e49b0835bbd3ac60baa685299e86777b064f/pkg/controller/template/render.go#L310-L357), where the support for `external` cloud providers will be added. For each platform with `external` configuration, the `cloud-config` flag will not be set.

> Kubelet is tightly coupled with `CCMs`, due to its [usage](https://github.com/kubernetes/kubernetes/blame/323f34858de18b862d43c40b2cced65ad8e24052/pkg/kubelet/cloudresource/cloud_request_manager.go#L98-L102) of cloud-provider interfaces, such as `Instances` and `NodeAdresses`. Whenever there is a need to bootstrap any `Node`, `kubelet` will collect missing information about the instance IP addresses, type and zone. This information could later be used by `kube-scheduler`.

Reviewer 6: collect from where? How do kubelet and external CCM interact?

Reviewer 6: What does "could later be used by kube-scheduler" mean? Is that a proposal for a change?

#### Kube Controller Manager and Kube API Server

Configuration of these components is managed by `cluster-kube-controller-manager-operator` and `cluster-kube-apiserver-operator` respectively. Both operators use [`configobserver`](https://github.com/openshift/library-go/blob/16d6830d0b80dc2a3315207116d009ed2dd4cebf/pkg/operator/configobserver/cloudprovider/observe_cloudprovider.go#L140) module from `library-go` where we set what cloud provider should be used for the given platform. When a platform is being switched to `CCM`, we [must not](https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/#running-cloud-controller-manager) set `cloudProvider` value there and leave it empty.

Next step is to bump `library-go` for the operators.

##### Pre-release / Development

In order to achieve desired cluster configuration for cloud components during development, we need to add an annotation on the `Infrastructure` [resource](https://github.com/openshift/api/blob/c2f7aea5d89ee87d7510b50bdb08dde9028a53ef/config/v1/types_infrastructure.go#L11).

Proposed annotation: `infrastructure.openshift.io/external-cloud-provider`.

> The annotation will determine the architectural change in the cluster configuration. If this annotation would be found, other components, such as `kube-apiserver`, `kube-controller-manager` and `kubelet` will set their flags accordingly, to allow hosting external `CCM` for any platform.

Reviewer 6: who will set the annotation? This flow with an exact timeline and actors must be much clearer.

Author: This annotation is set by us or CI at a running and non-upgrading cluster, to disable cloud provider flags across legacy components. An alternative could be using a set of pre-built images, a cluster will run, which we, as the developers will manually re-apply for every component. Both are temporary hacks, the first one is just making our lifes simpler.

Reviewer 6: my question was mainly about the upgrade timeline. Which operator will set the annotation at which point in time. 

Also consider downgrades.

Reviewer 3: Annotations are no longer included here, there will be a feature gate instead that users will optionally set. No operator should be setting this automatically.

The annotation will be a temporary measure to simplify the operator development allowing multiple components to merge implementations of this feature independently before switching on the feature by changing the value of the annotation. Once all providers are moved out-of-tree, the annotation will be safe to remove.

To ensure a smooth transition from the `in-tree` selection, `library-go` [implementation](https://github.com/openshift/library-go/blob/master/pkg/operator/configobserver/cloudprovider/observe_cloudprovider.go) will include an additional argument for the `cloudProviderObserver` structure, which will allow to preserve initial configuration strategy, until the new `observeExternal` flag will be set to `false`:

```go
type cloudProviderObserver struct {
    targetNamespaceName     string
    cloudProviderNamePath   []string
    cloudProviderConfigPath []string
    observeExternal         bool
}
```

> The `observeExternal` field will determine, when the annotation is present on the `Infrastructure` resource, it will be taken into account while establishing needed flags for a Kubernetes binaries. It will make the observer type to the `cloud-config` file, and set flags as was described in the `kubelet`, `kube-apiserver` and `kube-controller-manager` sections.

Reviewer 6: is `observeExternal` state of the controller? Is it an argument for the observer to behave differently? Why would we need that? Why not hardcoded (as @soltysh suggests) ?

Author: Dropping this info, it is an implementational detail. I still expect an annotation on the Infra resource to be used in development.

#### Bootstrap changes

> For the initial transition from in-tree to the out-of-tree, `CCM` pod will be created as static pod on the bootstrap node, to ensure swift removal of the `node.cloudprovider.kubernetes.io/uninitialized: true` taint from any newly created `Nodes`. Later stages, including cluster upgrades will be managed by an operator, which will ensure stability of the configuration, and will run a `CCM` in a `Deployment`. Initial usage of the static pod is justified by the need to run `CCM` pod before the `kube-apiserver`, `etcd` and the `kube-scheduler` will get healthy, so the functionality does not rely on them.

Reviewer 5: This is not quite correct, ccm can't run before etcd and kas, b/c that won't work. It's etcd and kas first, then ccm (which requires k8s api to be able to remove the taint from the node) then kcm, ks. But I'd just write that ccm is required at bootstrap to remove taints from the newly created nodes.

`CCM` operator will provide initial manifests that allow to deploy `CCM` on the bootstrap machine with `bootkube.sh` [script](https://github.com/openshift/installer/blob/master/data/data/bootstrap/files/usr/local/bin/bootkube.sh.template).

> At the initial stage of `CCM` installation installer creates a config map with `cloud.conf` key that contains configuration of `CCM` in `ini` format. The contents of `cloud.conf` are static and generated by the installer:

Reviewer 5: ini format is not relevant.

Example for OpenStack:

```txt
[Global]
secret-name = openstack-credentials
secret-namespace = kube-system
```

Resources deployed here will be destroyed with the bootstrap stage cleanup.

#### Operator resource management strategy

Every cloud manages it's configuration differently, but shares a common overall structure.

1. `cloud-controller-manager` and related resources are deployed under a single namespace.

> 2. Each pod (`CCM`, `cloud-node-manager`, etc.) will be managed by a `Deployment`.

Reviewer 5: what is cloud-node-manager? If you're talking about node controller that runs inside ccm, as far as I know.

Reviewer 5: I'm missing concrete information what are these each pod managed by deployment.

Reviewer 3: There are 3 or more controllers that normally make up CCM; Node, Service, Route and any additional features the provider wants to add.

Normally these all end up as one binary within CCM, but at least one of the providers (IIRC Azure) packages the CNM as a separate binary by default. My suggestion in this case would be to run this as a separate container within the same Pod rather than a separate Pod, but we could also look at running this as a single binary by building it in a different way within OpenShift.

@Danil-Grigorev Can we try to add a list of the providers and how they package this?

3. The pods are expected to run on the control-plane `Nodes` using `hostNetwork`.

Operator will manage:

- Dynamic provider resources, such as `Deployments`.
- Operator will populate, create and manage the cloud-specific set of `Deployments`, based on the selected `platform` in the `Infrastructure` resource.

> - `Deployments`, provisioned by the operator will run the pods on the `control-plane` nodes, and will tolerate `cloudprovider.kubernetes.io/uninitialized` and `node-role.kubernetes.io/master` taints. They will run at least `2` replicas of the `CCM` pods with anti-affinity settings to prevent scheduling on the same machine and allow some fault-tolerance.

Reviewer 6: I read Deployment*s* here. Why multiple?

Reviewer 3: I think this is a wording issue meant for the fact that there will be multiple different options for the deployment, one for AWS, one for Azure etc.

As far as I'm aware we are all in agreement that the operator should only create one deployment per cluster (@Danil-Grigorev @Fedosin  please thumbs up or show your agreement somehow)

> - The `--cloud-provider` and the `--cloud-config` arguments will be populated using existing `library-go` based implementation from the [cluster-kube-controller-manager-operator](https://github.com/openshift/cluster-kube-controller-manager-operator/blob/990ea3f2ace13be6579c00a889842c3f5d3e756a/pkg/operator/configobservation/configobservercontroller/observe_config_controller.go#L82-L85) to share a common approach with the bootstrap specific static-pod configuration.

Reviewer 5: I'm not quite sure what you mean with this statement. You've already described KCM-related changes sooner. Most probably you're again going into unnecessary details.

> - Operator will own a `cluster-operator` resource, which will report the readiness of all workloads at post-install phase. The conditions reported will help other dependent components, such as [openshift-ingress](https://github.com/openshift/cluster-ingress-operator).

Reviewer 5: How this will help other components? I don't think this is relevant.

Author: But the route provisioning in openshift is dependent on the CCM, and they would need to account for its readiness before starting any route serving. And this is common to have a `cluster-operator` for every operator, at least for basic status reporting. I might move it to it's own section, with only operator details.

Reviewer 3: Do we have something in this document explaining that route provisioning is dependent on this? We should do if we haven't already, dependencies need to be understood

Reviewer 5: Each component in the OCP cluster should be capable of reacting to failure of dependants. I agree this should be called out in the enhancement for clarity.

- Operator will be built with consideration for potential future implementations of hybrid OpenShift clusters which support multiple cloud providers simultaneously (ie ensure resources created for each provider by the operator do not clash with resources for other providers)

*Note: static resource creation is delegated to CVO (SA, RBAC, etc.)* 

#### CVO managemnet

The operator image will be build and included in every release payload, and expected to be deployed and running at kubernetes operators level ([10-29](https://github.com/openshift/cluster-version-operator/blob/34010292c3abd2582b687e9c0ef76c5924998f39/docs/dev/operators.md#how-do-i-get-added-as-a-special-run-level)) right after the `kube-controller-manager-operator` at runlevel [25](https://github.com/openshift/cluster-kube-controller-manager-operator/tree/master/manifests).

Proposed `CCMO` manifests runlevel is `26`.

#### Cloud provider management

> Each cloud provider will be hosted under its own repository. Those repositories will be initially bootstrapped by forking upstream ones.

Reviewer 5: Why forking? What changes are needed that you need to fork from the start? I'd imagine us using upstream wherever possible.

Reviewer 3: Is it not standard in Red Hat to fork a project into one of our orgs so that we can set up our own CI and build/release automation? Like we do for the cluster-autoscaler for example.

Is there some other way we can build the artifacts required without forking?

Author: We will try to be but in order to preserve openshift requirements on security, we might want to use a different set of default ports, services, containers for each provider.

Author: We also need to support a couple of releases down the line, have a custom Dockerfile and a well-known QA coverage for the CCM code, before putting it into the release.

Reviewer 5: > We will try to be but in order to preserve openshift requirements on security, we might want to use a different set of default ports, services, containers for each provider.

This is configuration done by the operator, it should not require fork.

Dockerfile and RH built-related stuff are reasonable, though.

The `CVO` will be responsible for provisioning static resources, such as `ServiceAccounts`, `Roles/RoleBindings`, `ServiceMonitors` for metrics for the operator. Then, the operator will  be responsible for constructing and creating cloud provider resources.

Forked provider repositories will also contain cloud-provider specific code, and a set of binaries for each of them to build and run. Each provider will be build inside a single image, which will be included in the release payload. The responsibility for the operator will be to choose which provider image should be used in a cluster at any moment.

#### CCM migration from KCM pod

Currently the cloud-specific controller loops are running inside the `kcm` pod. We need to preserve leadership over these parts during migration process. This includes migration from the `openshift-kube-controller-manager` namespace to the `openshift-cloud-controller-manager` namespace, which is not possible with current `leader-election` implementation.

> Upstream [proposal](https://github.com/kubernetes/enhancements/blob/473d6a094f07e399079d1ce4698d6e52ddcc1567/keps/sig-cloud-provider/20190422-cloud-controller-manager-migration.md#motivation) describes the main implementation details for `Lease` resource. This cluster-scoped resource maintains cross-namespaced component leadership, and would help OpenShift to preserve HA during migration from the in-tree to the out-of-tree.

Reviewer 6: what does it mean? We add a cross-ns lease mechanism to kcmo or to kcm?

> As an alternative to this implementation, the `CCCMO` could be responsible for reading the state of the `kcm` cluster-operator status, and waiting for `Avaliable` status to be `true`, and `Progressing` to become `false`, which would mean that all `nodes` are at a new revision of the `kcm` with the `CCM` controllers disabled (or enabled, depending on the provider and release version). Then the `CCCMO` will start provisioning nessesary `CCM` resources.

Reviewer 5: You shouldn't wait for other operator availability. Whoever is first "wins" the lease and is the main process. The way the migration is described it's possible for either of the controller-managers to grab the lock and run.

Reviewer 6: Would kcm give up its lease eventually?

Reviewer 3: This suggestion came from the upstream docs but I don't think it's been worded very clearly here.

The idea is that you: 
- Update KCM to turn off the controllers that will be taken over by CCM.
- Wait for KCM to deploy and become ready, ie, old pods have gone away, new pods are not running the CCM controllers
- Now you can deploy CCM knowing that nothing will conflict

I think the idea was that we could build this logic into the CCCMO so that it could only create the deployment for the CCM once it's sure that KCM is no longer running the cloud controllers. This would involve a small amount of downtime for the cloud controllers though, is that acceptable?

Reviewer 5: It would be good to add just that to this proposal, so anyone looking at it has a clear idea how the migration will look like.

*Alternative approach (long-term)*

An upstream effort in this [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cloud-provider/20191004-out-of-tree-credential-providers.md) attempts to remove `Kubelet` dependency on in-tree credentials issuing process, and instead use a similar approach with pluggable credentials providers. This interface is supposed to be moved into `Alpha` stage with this [PR](https://github.com/kubernetes/kubernetes/pull/94196) in `1.20` release.

### Upgrade/Downgrade strategy

A set of `cloud-controller-manager` pods will be running under a `Deployment` resource, provisioned by the operator. The leader-election will preserve the leadership during the cluster updates over cloud-specific loops.

Upgrade from previous versions of OpenShift on OpenStack will look like:

#### 4.6 -> 4.7

Upgrade (all platforms):

- The `cluster-version-operator` starts `cluster-cloud-controller-manager-operator`.

Downgrade (all platforms):

> - `cluster-cloud-controller-manager-operator` de-provisions CCM `deployments` and stops working.

Reviewer 6: in this section I would expect a discussion about the lifecycle of the annotation, and about the leases.

Author: The featureGate implementation will cover the acquire or loose lease procedure for lease between `kcm` and `ccm`.

The steps below apply to OpenStack only:

Upgrade:

- The `cluster-version-operator` starts `cluster-cloud-controller-manager-operator`.

> - `cluster-cloud-controller-manager-operator` creates all required resources for CCM (RBAC, Service Account, etc.).

Reviewer 5: Details, I'm looking for openstack specific details. The flow as is presented is generic and could be written above, since as I understand it, it applies to all platforms

For openstack specific section I'd like to see ALL the compoments the operator will deploy for OpenStack. How many, which deployments, are all only deployments, other? That's the kind of details missing for OpenStack.

> - `kube-apiserver` and `kube-controller-manager` are restarted without the `--cloud-provider` option.

Reviewer 5: kas-o and kcm-o restart kas and kcm updating --cloud-provider option

> - `kubelet` is restarted with `--cloud-provider=external` option.

Reviewer 5: mco restarts kubelet updating cloud-provider

Downgrade:

- `cluster-cloud-controller-manager-operator` de-provisions CCM `deployments` and stops working.
- `kube-apiserver` and `kube-controller-manager` are restarted with the `--cloud-provider=<some-platform>` option.

> - An old machine configuration is applied, which causes `kubelet` to restart with `--cloud-provider=<some-platform>` option.

Reviewer 5: Similarly, write who restarts

#### 4.7 -> 4.8

Upgrade:

> - `cluster-cloud-controller-manager-operator` creates resources for any in-tree provider defaulting to this in 4.8 (expecting: AWS, Azure, GCP, vSphere).

Reviewer 5: How this is different than the previous one? I'd like to see a timeline (can be a proposal) when each provider will be migrated to out-of-tree. Example:
- 4.7: GCP
- 4.8: AWS
- 4.9: all
etc.

Author: The release for each provider is going to depend on CSI plugin and migration readiness. Once both are shipped with the release, out-of-tree implementation will follow. The upper boundary is the exclusion of the in-tree upstream (starting in 1.23 for kubelet as of now) - appr 4.10.

- `kubelet` is restarted with `--cloud-provider=external` option.

Downgrade:

- `cluster-cloud-controller-manager-operator` de-provisions CCM resources for any platform, except OpenStack.
- `kube-apiserver` and `kube-controller-manager` are restarted with the `--cloud-provider=<some-platform>` option.
- An old machine configuration is applied, which causes `kubelet` to restart with `--cloud-provider=<some-platform>` option (excluding OpenStack).

#### 4.8 and later

Upgrade:

- `cluster-cloud-controller-manager-operator` updates the state of resources for any in-tree provider defaulting to this since previous releases.

Downgrade:

- `cluster-cloud-controller-manager-operator` downgrades the state of resources for any in-tree provider defaulting to this since previous releases.

### Version Skew Strategy

See the upgrade/downgrade strategy.

### Action plan (for OpenStack)

#### Build OpenStack CCM image by OpenShift automation

To start using OpenStack CCM in OpenShift we need to build its image and make sure it is a part of the OpenShift release image. The CCM image should be automatically tested before it becomes available.

> The upstream repo provides the Dockerfile, so we can reuse it to complete the task. Upstream image is already available in [Dockerhub](https://hub.docker.com/r/k8scloudprovider/openstack-cloud-controller-manager).

Reviewer 5: We will built our own images.

Actions:

- Configure CI operator to build OpenStack CCM image.

> CI operator will run containerized and End-to-End tests and also push the resulting image in the OpenShift Quay account.

Reviewer 5: What kind of e2e you're expecting to run, I'd like at least high level overview of what's going to be tested.

#### Test Cloud Controller Manager manually

> When all required components are built, we can manually deploy OpenStack CCM and test how it works.

Reviewer 5: Manual is ok for dev testing, we need a proper set of automatic tests ensuring full functionality of the operator and its operands.

Reviewer 9: agree! now we don't have a dedicated openstack-upgrade ci job, but it would be nice to have one to reduce migration risks and minimize the number of potential errors.

Actions:

- Manually install CCM's example [daemonset](https://github.com/kubernetes/cloud-provider-openstack/blob/master/manifests/controller-manager/openstack-cloud-controller-manager-ds.yaml) on a working OpenShift cluster deployed on OpenStack.

- Update configuration of `kubelet` by replacing `--cloud-provider openstack` with `--cloud-provider external` and removing `--cloud-config` parameters.
For `kube-apiserver` and `kube-controller-manager` we need to remove both `--cloud-provider` and `--cloud-config` parameters and restart `kubelet`.

- Check the CCM functionality is operational at this point.

**Note:** Example of a manual testing: https://asciinema.org/a/303399?speed=2

> #### Write Cluster Cloud Controller Manager Operator

Reviewer 5: This and next section are not relevant.

The operator should be able to create, configure and manage different CCMs (including one for OpenStack) in OpenShift. The architecture of the operator will be following practices from [Kubernetes Controller Manager](https://github.com/openshift/cluster-kube-controller-manager-operator) and [Kubernetes API server](https://github.com/openshift/cluster-kube-apiserver-operator) operators.

Actions:

- Create a new repo in OpenShift’s github: https://github.com/openshift/cluster-cloud-controller-manager-operator (Done)

- Implement the operator, using [library-go](https://github.com/openshift/library-go) primitives.

#### Build the operator image by OpenShift automation

To start using the CCM operator in OpenShift we need to build its image and make sure it is a part of the OpenShift release image. The image should be automatically tested before it becomes available. Dockerfile should be a part of the operator's repo.

Actions:

- Configure CI operator to build CCM operator image.

CI operator will run containerized and End-to-End tests and also push the resulting image in the OpenShift Quay account.

> #### Integrate the solution with OpenShift

Reviewer 5: You already said that, not relevant.

#### Examples for other cloud-controller-manager configurations

- AWS: https://github.com/kubernetes/cloud-provider-aws/
- Azure: https://github.com/kubernetes-sigs/cloud-provider-azure
Azure separated the `CCM` and the "cloud-node-manager" (`CNM`) into two binaries: https://github.com/kubernetes-sigs/cloud-provider-azure/tree/master/cmd.
- GCP: https://github.com/kubernetes/cloud-provider-gcp
- vSphere: https://github.com/kubernetes/cloud-provider-vsphere
vSphere introduced an additional service to expose internal API server: https://github.com/kubernetes/cloud-provider-vsphere/blob/master/manifests/controller-manager/vsphere-cloud-controller-manager-ds.yaml#L64

### Risks and Mitigations

#### Specific CCM doesn’t work properly on OpenShift

The `CCMs` have not been tested on OCP yet. Unknown issues may arise.

Severity: medium-high
Likelihood: medium

#### OpenShift components can't work with the CCM properly

Since OpenShift imposes additional limitations compared to Kubernetes, some collisions are possible.
So far there are no CCMs available in OpenShift, and there can be some problems with operability.

Severity: medium-high
Likelihood: low

> #### CSI driver is not available in OpenShift

Reviewer 5: I have no idea how CSI drivers fall in the big picture of out-of-tree providers, it wasn't covered earlier.

Reviewer 3: Does the updated content below help or should we be mentioning this earlier in the proposal?

CCM doesn't support the in-tree `Volume` manager implementation. A CSI driver now becomes a mandatory requirement for every out-of-tree provider. Additionally the CSI should support data migration between the in-tree storage solution and the plugin provided storage interfaces.

Severity: medium-high
Likelihood: low

### Test Plan

- Make sure the `CCM` implementation for "any" cloud is up and running in a bootstrap phase before or shortly after the `kubelet` is running on any `Node`.
- Ensure that transition between the in-tree and the out-of-tree will be handled by the established architecture with the new release payload, where all of the existing `Nodes` are operational during and after the upgrade time, and the bootstrap procedure for new ones is successful.
- Ensure that upgrade from a release running on the in-tree provider only to the out-of-tree succeeds, and the downgrade is also supported.
- Each provider should have a working CI implementation, which will assist in testing the provisioning and upgrades with out-of-tree support enabled.

### Graduation Criteria

#### Tech preview

> Operator handles OpenStack installation in 4.7.

Reviewer 6: this surprises me. Does this mean we only switch to external for OpenShift as opt-in i.e. optionally?

Author: Mike suggests some optional environment variable for the installer to recognize it. Could you please elaborate, @Fedosin?

#### Tech Preview -> GA

The operator will handle basic tasks, such as create initial templates at bootstrap and running all of supported in-tree providers after installation.

## FAQ

Q: Can we disaster-recover a cluster without CCM running?

> A: We can't. Basically, without CCM running, nodes can only join the cluster, but they will be unschedulable. Which means it's impossible to start any workloads on the nodes. [Source](https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/#running-cloud-controller-manager)

Reviewer 5: If that's the case, how are you planning to upgrade the current disaster recovery mechanism to include ccm in it?

Reviewer 5: This is not covered in the implementation section, at all.

Reviewer 3: Do you have a link to something that explains the current DR mechanism? Would be good to be able to refer to this as we explain the changes that are required.

Reviewer 5: https://github.com/openshift/enhancements/blob/master/enhancements/etcd/disaster-recovery-with-ceo.md is for etcd, https://github.com/openshift/enhancements/blob/master/enhancements/kube-apiserver/auto-cert-recovery.md is for certs

Q: What is non-functional in a cluster during bootstrapping until CCM is up?
A: If CCM is not available, new nodes in the cluster will be left unschedulable.

Q: Who talks to CCM?

> A: From an architectural point of view, only Kube API server is able to communicate with CCM. [Source](https://kubernetes.io/docs/concepts/architecture/cloud-controller/#design). All requests to CCM are sent by `kubelet`. KCM doesn't interact with CCM.

Reviewer 5: That picture is wrong, ccm talks to kas, to update taints on node, for starters. What about kubelet and kcm?

Reviewer 5: Please open an upstream PR fixing this picture after you identify the correct order of the arrows on that picture.

Reviewer 3: Agreed this is wrong and needs an update.

In the upstream picture they have it as if KA talks to KCM and KS, which is also wrong, the arrow from KA to CCM is consistent with this though

A: A couple of components rely on CCM at different time of the cluster lifetime:
  - `kubelet` dependency requires removal of taints from newly created Nodes.
  - CCMs manage `LoadBalancer` `Service` resources in all in-tree cloud implementations, except vSphere, so their creation will not be possible while the CCM is down. This will affect the [ingress-operator](https://github.com/openshift/cluster-ingress-operator). Cloud networking across `Nodes` is [disabled](https://github.com/openshift/cluster-kube-controller-manager-operator/blob/8be94db9da523152af2268bd7f891fe089a424eb/bindata/v4.1.0/config/defaultconfig.yaml#L8-L9) by default for all clouds, so the routes will not be provisioned for the Node by the `CCM` (which is done by `Routes()` method in the interface), and so not required for SDN functionality.
  - Storage CSI migration will be required for OpenStack and vSphere providers

Q: Should we reuse the existing cloud provider config or generate a new one?
A: CCM config is backward compatible with the in-tree cloud provider. It means we can reuse it.

Q: How to migrate PVs created by the in-tree cloud provider?

> A: [CSIMigration](https://kubernetes.io/blog/2019/12/09/kubernetes-1-17-feature-csi-migration-beta/) looks like the best option, especially if it is GA in 1.19.

Reviewer 6: Volumes are migrated? What is the downgrade strategy for this step?

Reviewer 10: 1. CSI migration most probably won't GA in 1.19 upstream.
2. Storage team does not plan to support it in OCP 4.7, not even as tech preview. Therefore, if an existing cluster is upgraded to external CCM, it looses access to *all* its PVs. For freshly installed clusters, making the CSI storage class as default looks feasible in 4.7 timeframe.

Q: Does CCM provide an API? How is it hosted? HTTPS endpoint? Through load balancer?

> A: No. The CCM vendors part of the code serving cloud features from the [cloud-provider](https://github.com/kubernetes/cloud-provider). CCM provides a secure TCP connection on port [10258](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cloud-provider/ports.go#L22) and communicates similarly to other kubernets components.

Reviewer 5: So you do expose an endpoint, what it does?

Author: This information is wrong. This endpoint is used only for `healthz` and `readyz` and other services describing the component status, like metrics. All communication is done in the Node resource, no component is directly talking to the `CCM`. The out-of-tree CCM is vendoring the native cloud-provider code, no change from what `kube-controller-manager` did. It is just a default port, going to remove it.

Reviewer 5: Right, but you are exposing an endpoint, even though it's for healthz and readyz it should be called out b/c it's an important piece of information for people looking at internal cluster communication.

Reviewer 3: This has been updated in the last push and is now much more clear

Q: What happens if the kcm leader has no access to ccm? Can it continue its work? Will it give up leadership?
A: Expect KCM perform independently to CCM and do not care about its state after initial code migration. The CCM cluster operator will be responsible for going Unavailable or Degraded in a similar situation.

Q: What are the thoughts about certificate management?

> A: Certificate management should be performed by the CCM operator, like it's done in the KCM operator.

Reviewer 5: What kind of certificate management?

Author: GCP has the logic related to certificate management. It approves some CSR requests: https://github.com/kubernetes/cloud-provider-gcp/blob/master/cmd/gcp-controller-manager/csr_approver.go#L87

Reviewer 5: How this works in combination with OCP's certificate auto-rotation?

Reviewer 3: I think we are talking about two different things here, there's some certificate management for Node certs mentioned with respect to GCP, but that's been resolved within the latest update.

What we haven't touched, which I think is @soltysh's point, is that the CCM pod will need certificates to host the metrics/health endpoints, and we haven't discussed how that will be tackled, we need to address this

Q: What happens if the KCM leader has no access to CCM? Can it continue its work? Will it give up leadership?
A: Since KCM does not communicate with CCM it can continue to work if CCM is not available.

Q: Does every node need a CCM?

> A: No. Control plane nodes only. [Source](https://kubernetes.io/docs/concepts/overview/components/#cloud-controller-manager)

Reviewer 5: All? Earlier you've stated that you will run 2 replica deployment.

Author: It is safe to say control-plane nodes, as they are the primary consumers for CCM and new Node creation process, and the most stable ones (workers could go away any moment). There should be some fault tolerance, so ~2 pods located on different Nodes is disired.

Reviewer 5: All of our core components run on master nodes, always. We always run 3 pods for HA. With 2 pods you'll be having issues deciding which of the 2 is the current leader and manages CCM-related operation. Look at how kas, kcm or ks are deployed, since your deployment should match that.

Reviewer 3: I don't think there should be any issue with running 2 if we use the leader election implementation from controller-runtime or client-go ([example](https://github.com/kubernetes/client-go/blob/master/examples/leader-election/main.go)), the mechanism is just a first to grab the lock kind of thing which uses KAS as the arbiter of conflicts. As far as I'm aware this mechanism is used across OpenShift already and doesn't need 3 replicas, there's no raft style communication between different potential leaders, that's deffered to etcd.

We can make this 3 replicas if that is consistent with KAS/KCM but I don't feel like this is necessary

Q: How metrics are affected by the CCM migration?

> A: Currently the `CCM` implementation is using the same set of metrics from the core repository. The `CCMO` will require prometheus configuration to collect them from a different location.

Reviewer 5: What kind of metrics are exposed, which are relevant for cluster administrator

Author: - https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/legacy-cloud-providers/gce/metrics.go

Reviewer 3: It was my understanding that the metrics that are being exposed haven't changed, they just need to come from a new place, since this isn't a change, do we really need to document what the particular metrics are?

Reviewer 5: I think high-level overview of what we provide would be desirable.

Author: https://github.com/openshift/enhancements/pull/463/files#diff-3e0e2c48e70215076dfe36c13768a823ab7080d929d80292f37db2ef5a2121e8R231

## Other notes

### Removing a deprecated feature

> Upstream is currently planning to remove support for the in-tree providers with OCP 4.8 (kubernetes 1.21) - based on the slack [conversation](https://kubernetes.slack.com/archives/CPPU7NY8Y/p1589289501020600). The enhancement [describes](https://github.com/kubernetes/enhancements/blob/473d6a094f07e399079d1ce4698d6e52ddcc1567/keps/sig-cloud-provider/20190125-removing-in-tree-providers.md) steps towards this process. Our goal is to have a working alternative for the in-tree providers ready to be used on-par with old implementation before the upstream release will remove this feature from core Kubernetes repository.

Reviewer 5: If you're talking about upstream revert order, k8s first and ocp in parents.

- Operator will populate, create and manage the cloud-specific set of `Deployments`, based on the selected `platform` in the `Infrastructure` resource.
- `Deployments`, provisioned by the operator will run the pods on the `control-plane` nodes, and will tolerate `cloudprovider.kubernetes.io/uninitialized` and `node-role.kubernetes.io/master` taints.
- The `--cloud-provider` and the `--cloud-config` arguments will be populated using existing `library-go` based implementation from the [cluster-kube-controller-manager-operator](https://github.com/openshift/cluster-kube-controller-manager-operator/blob/990ea3f2ace13be6579c00a889842c3f5d3e756a/pkg/operator/configobservation/configobservercontroller/observe_config_controller.go#L82-L85) to share a common approach with the bootstrap specific static-pod configuration.
- Operator will own a `cluster-operator` resource, which will report the readiness of all workloads at post-install phase. The conditions reported will help other dependent components, such as [openshift-ingress](https://github.com/openshift/cluster-ingress-operator).

> - Operator will be built with a goal to achieve provider decoupling and potential switching between providers in the cluster, and will consider support to run multiple providers in a cluster simultaneously (long-term goal).

Reviewer 8: Not sure what `switching between providers` mean here.

At first I understood you meant deploying on one platform, eg AWS, and later migrate your cluster to another platform such as OpenStack. It would be fantastic but this looks like science fiction to me and is way out of scope for this enhancement.

I assume you meant switching between different implementations of the same CCM?

Reviewer 3: We could drop this, or reword, I've added a suggestion below which I think better explains how I had understood what this point was going for
```suggestion
- Operator will be built with consideration for potential future implementations of hybrid OpenShift clusters which support multiple cloud providers simultaneously (ie ensure resources created for each provider by the operator do not clash with resources for other providers)
```

#### CVO managemnet

> The operator image will be build and included in every release payload, and expected to be deployed and running at kuberentes operators level ([10-29](https://github.com/openshift/cluster-version-operator/blob/34010292c3abd2582b687e9c0ef76c5924998f39/docs/dev/operators.md#how-do-i-get-added-as-a-special-run-level)) right after the `kube-controller-manager-operator` at runlevel [25](https://github.com/openshift/cluster-kube-controller-manager-operator/tree/master/manifests).

Reviewer 8: s/kuberentes/kubernetes/

## Summary

This enhancement proposal describes the migration of cloud platforms from the deprecated [in-tree cloud providers](https://v1-18.docs.kubernetes.io/docs/concepts/cluster-administration/cloud-providers/) to [Cloud Controller Manager](https://kubernetes.io/docs/concepts/architecture/cloud-controller/) services that implement the`external cloud provider` [interface](https://github.com/kubernetes/cloud-provider/blob/master/cloud.go).

## Motivation

Using Cloud Controller Managers (CCMs) is the Kubernetes [preferred way](https://kubernetes.io/blog/2019/04/17/the-future-of-cloud-providers-in-kubernetes/) to interact with the underlying cloud platform as it provides more flexibility and freedom for developers.
It replaces existing in-tree cloud providers, which have been deprecated and will be permanently removed approximately in Kubernetes `v1.23`.
In-Tree providers are still used within OpenShift and we must start a smooth migration towards CCMs.

Another motivation is to be closer to upstream by helping develop Cloud Controller Managers for various platforms, which is benefiting both OpenShift and Kubernetes.

> The change will help adding support for other cloud platforms, such as [Packet](https://github.com/packethost/packet-ccm), [Digital Ocean](https://github.com/digitalocean/digitalocean-cloud-controller-manager) or [Alibaba Cloud](https://github.com/kubernetes/cloud-provider-alibaba-cloud).

Author: There is an open enhancement for the Packet - https://github.com/openshift/enhancements/pull/436 and a couple of jiras ( CORS-1273, OCPPLAN-1359)

> The change will help adding support for other cloud platforms, such as [Packet](https://github.com/packethost/packet-ccm), [Digital Ocean](https://github.com/digitalocean/digitalocean-cloud-controller-manager) or [Alibaba Cloud](https://github.com/kubernetes/cloud-provider-alibaba-cloud).

Reviewer 3: The Packet provider for OCP is in the process of being built out by a team from Packet

#### CCM migration from KCM pod

Currently the cloud-specific controller loops are running inside the `kcm` pod. It is ok to run in-tree and out-of-tree `CCM` simultaniously in a cluster for a short amount of time, or have a short amount of downtime between the moment in-tree plugin shuts down, and out-of-tree starts, but this gap should not be experienced by components, which rely on `CCM` functionality. Those are:
- `opneshift-ingress` with `Route` resource management - non-operational without `CCM`.
- `kubelet` - can't initialize a new `Node` without `CCM`.

Although `CCM` controller loops itself does not require HA setup, the in-tree storage plugins are removed together with the `CCM`. This results in `CCM` dependency on the [cluster-storage-operator](https://github.com/openshift/cluster-storage-operator), and it's ability to support the provider CSI plugin and perform migration from in-tree storage plugin to the CSI one.

#### Credentials management

> The operator is expected to be integrated with the [cloud-credentials-operator](https://github.com/openshift/cloud-credential-operator), and request fine grained credentials for the cloud components.

Reviewer 3: This will need rewording but I think that was always our intention. CVO will create a number of `CredentialsRequests` and the CCO will provision appropriately. I think this is described below in `Post-install phase`

#### Pre-release / Development Feature Gate

In order to achieve desired cluster configuration for cloud components during development and technical preview, we will add a new [feature gate](https://docs.openshift.com/container-platform/4.5/nodes/clusters/nodes-cluster-enabling-features.html) `ExternalCloudProvider`, which will allow developers and customers to force their cluster to upgrade from in-tree to out-of-tree cloud providers.
By using this feature gate, users will prevent their clusters from upgrading to newer versions of OpenShift as with all other feature gates.

The feature gate will be used to determine whether or not operators should migrate their components to the out-of-tree configuration.
If the feature gate  is listed (via a custom or technical preview feature set), the operators for components such as `kube-apiserver`, `kube-controller-manager` and `kubelet` will update their component's configuration accordingly to allow hosting external `CCM` for any platform.

Once an out-of-tree provider implementation is ready to be released (GA), it will no longer depend on the feature gate and will automatically upgrade as part of a 4.N to 4.N+1 upgrade.

> Once all providers are moved out-of-tree, the feature gate will be safe to remove.

Reviewer 3: @soltysh @sttts Based on some input last week we have reworded this section to remove all reference to an annotation and are now planning to use a feature gate instead. Would appreciate your eyes on this description to make sure this makes sense and see if there's anything else you'd like us to document about this

Q: CCM relation to other openshift components, such as SDN and storage? How a non-operational CCM will affect cluster health, which components will take a hit?
A: A couple of components rely on CCM at different time of the cluster lifetime:
  - `kubelet` dependency requires removal of taints from newly created Nodes. Without CCM, new nodes will not become schedulable, existing nodes are unaffected.
  - CCMs manage `LoadBalancer` `Service` resources in all in-tree cloud implementations, except vSphere, so their creation will not be possible while the CCM is down. This will affect the [ingress-operator](https://github.com/openshift/cluster-ingress-operator). Cloud networking across `Nodes` is [disabled](https://github.com/openshift/cluster-kube-controller-manager-operator/blob/8be94db9da523152af2268bd7f891fe089a424eb/bindata/v4.1.0/config/defaultconfig.yaml#L8-L9) by default for all clouds, so the routes will not be provisioned for the Node by the `CCM` (which is done by `Routes()` method in the interface), and so not required for SDN functionality.

>   - Storage CSI migration will be required for OpenStack and vSphere providers

Reviewer 11: Also for AWS, GCE and Azure, no?

## Other notes

### Removing a deprecated feature

Upstream is currently planning to remove support for the in-tree providers with kubernetes 1.23. The enhancement [describes](https://github.com/kubernetes/enhancements/blob/473d6a094f07e399079d1ce4698d6e52ddcc1567/keps/sig-cloud-provider/20190125-removing-in-tree-providers.md) steps towards this process. Our goal is to have a working alternative for the in-tree providers ready to be used on-par with the old implementation before upstream remove this feature from core Kubernetes repository.

The `cloud-provider` and `cloud-config` flags from `kubelet` will be removed in 1.23 - [link](https://github.com/kubernetes/kubernetes/pull/90408).

## Timeline

### Upstream

Follow the Kubernetes community discussions and implementation/releases

- vSphere support has graduated to beta in [v1.18](https://github.com/kubernetes/enhancements/issues/670) (Openshift 4.5)

- Azure support goes into beta in [v1.20](https://github.com/kubernetes/enhancements/issues/667) (Openshift 4.7)

### OpenShift

#### OCP 4.7

> - OpenStack platform on out-of-tree `CCM` graduates as tech-preview.

Reviewer 11: hmm, are you sure this is feasible? 

To be tech preview in 4.7, we need at least tech preview support for OpenStack CSI migration; and the Storage Team doesn't have anything planned for 4.7, except a sketch of a few (bad) ideas: https://github.com/bertinatto/enhancements-1/commit/af844363666331e9124573bf26f55da1ac0699ac

Reviewer 9: This patch should allow us to enable csi migration for testing: https://github.com/openshift/installer/pull/4304

Reviewer 3: I think the idea here, @Fedosin can confirm, was that we would have a very big warning that says that migration is not included and existing volumes will not be migrated. The intention being that users would only be using this feature on temporary testing clusters and not production systems where storage is a requirement.

We should expand this section to include this detail

#### OCP 4.8

> - AWS on out-of-tree `CCM` tech-preview. `CSI` is enabled since [4.5](https://github.com/openshift/enhancements/blob/master/enhancements/storage/csi-driver-install.md#ocp-45-kubernetes-118).

Reviewer 11: nit: it's actually 4.6. The 4.5 AWS EBS CSI Driver operator is not installed by default on every cluster as it's managed by OLM.

Reviewer 3: ```suggestion
- AWS on out-of-tree `CCM` tech-preview. `CSI` is optional (via OLM) since [4.6](https://github.com/openshift/enhancements/blob/master/enhancements/storage/csi-driver-install.md#ocp-45-kubernetes-118).
```

In OpenShift, `kube-controller-manager` and `kube-apiserver` are managed by `cluster-kube-controller-manager-operator` and `cluster-kube-apiserver-operator` respectively.
When switching to out-of-tree providers, the `cloud-provider` option [should be removed](https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/#running-cloud-controller-manager) from their deployments.

#### Pre-release / Development Feature Gate

In order to achieve desired cluster configuration for cloud components during development and technical preview, we will add a new [feature gate](https://docs.openshift.com/container-platform/4.5/nodes/clusters/nodes-cluster-enabling-features.html) `ExternalCloudProvider`, which will allow developers and customers to force their cluster to upgrade from in-tree to out-of-tree cloud providers.

By using this feature gate with `TechPreviewNoUpgrade` `FeatureSet`, users will prevent their clusters from upgrading to newer versions of OpenShift as with all other feature gates.

The feature gate will be used to determine whether or not operators should migrate their components to the out-of-tree configuration.

> If the feature gate  is listed (via a custom or technical preview feature set), the operators for components such as `kube-apiserver`, `kube-controller-manager` and `kubelet` will update their component's configuration accordingly to allow hosting external `CCM` for any platform.

Reviewer 6: what does "for any platform" mean? I assume not all platforms support CCM, or do they?

> Once an out-of-tree provider implementation is ready to be released (GA), it will no longer depend on the `TechPreviewNoUpgrade` feature gate and will automatically upgrade as part of a 4.N to 4.N+1 upgrade using default feature set.

Reviewer 6: will we support downgrade too?

Once all providers are moved out-of-tree, the feature gate will be safe to remove.

This feature gate will be commonly used in CI tasks, to ensure upgrade stability. It will also be used in conjunction with `CSIMigration*` feature gates, to start volume migration during upgrades. The implementation will require changes in [installer](https://github.com/openshift/installer) to support feature gates at install time.

#### Bootstrap changes

Once an out-of-tree provider is released (GA), the `CCM` will be created as a static pod on the bootstrap node, to ensure swift removal of the `node.cloudprovider.kubernetes.io/uninitialized: true` taint from any newly created `Nodes`. Later stages, including cluster upgrades will be managed by an operator, which will ensure stability of the configuration, and will run the `CCM` in a `Deployment`.
Initial usage of the static pod is justified by the need for `CCM` to initialise `Nodes` before the `kube-scheduler` is able to schedule regular worklods (eg the operator for `CCM`).

The order will be the following:

1) Start `etcd` and `kube-api-server`.

> 2) Run the `CCM`, which will be provisioned as a static pod.

Reviewer 6: static pod *on bootstrap node via render step*

> 3) Start the `kube-scheduler`.

Reviewer 6: there is no order afaik for ka, kcm, ks. All are copied into /etc/kubenetes manifests dir and kubelet will start them.

`CCCMO` will provide a way to generate initial manifests that allow deployment of `CCM` as a static pod on the bootstrap machine.

> This will be used with the `bootkube.sh` [script](https://github.com/openshift/installer/blob/master/data/data/bootstrap/files/usr/local/bin/bootkube.sh.template).

Reviewer 6: please use the "render command" style as the other operators.

At the initial stage of `CCM` installation, the installer creates a `ConfigMap` with a `cloud.conf` key that contains configuration of `CCM`.
The contents of `cloud.conf` are static and generated by the installer:

Example for OpenStack:

```txt
[Global]
secret-name = openstack-credentials
secret-namespace = kube-system
```

Resources deployed here will be destroyed by the bootstrap cleanup process.

> 

Reviewer 6: I miss something about the feature gate topic here in the section. The installer allows to set them? How does the bootstrap phase know? There is no feature gate CR yet.

Reviewer 3: This section is describing the post GA state, when that happens the conditional blocking on the feature gate will be removed and components will deploy as out of tree during bootstrap. The top of this section has been updated to include a note that this is for GA, I think this is resolved

### Upgrade/Downgrade strategy

A set of `cloud-controller-manager` pods will be running under a `Deployment` resource, provisioned by the operator. The leader-election will preserve the leadership during the cluster updates over cloud-specific loops.

Upgrades from previous versions of OpenShift will look like:

#### 4.6 -> 4.7

Upgrade (all platforms):

- The `cluster-version-operator` starts `cluster-cloud-controller-manager-operator`.
- Users may choose to add a feature gate to enable the out-of-tree provider for OpenStack.

Downgrade (any `CCM` supported platform):

- No-Op:
    - CVO cannot remove the new operator. It will sit dormant on a downgraded cluster.
    - If a user enabled the out-of-tree provider for OpenStack via the feature gate, this cannot be undone.

*OpenStack CCM will be in a tech-preview at 4.7, only a fresh cluster installation or users willing to enable the feature gate could optionally host out-of-tree CCM for OpenStack*

#### Post 4.7 (in-tree to out-of-tree per provider)

Each provider will be released independently as and when the providers are deemed stable enough.

Prerequisites for upgrade:

> - The provider must support `CSI` for storage.

Reviewer 6: "the provider in the target version"

- The provider must support `CSI Migration` from in-tree storage. `CSI Migration` must be approved (tests, CI, GA) for the provider upgrade by default.
- `CSI` `StorageClass` should be able to become default after upgrade.

Upgrade:

> - New `cluster-kube-apiserver-operator` and `cluster-kube-controller-manager-operator` version updates the `kube-apiserver` and the `kube-controller-manager` and removes the `--cloud-provider` flag.

Reviewer 6: how do they know for which cloud providers? The text above suggests that not all providers are ready in 4.8.

- `CSI Migration` is started, as described in [proposal](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/csi-migration.md#upgradedowngrade-migrateunmigrate-scenarios).

> - `cluster-cloud-controller-manager-operator` starts and creates out-of-tree `CCM` resources for the provider.

Reviewer 6: does there have to be any sync between these steps?

Imagine ka, kcm, ks start with ccm support, but CSI migration is not finished or CCM is not up. What does this mean?

Reviewer 3: This will be kept in sync by the migration mechanism above which will have to exist in all openshift versions until all providers are migrated. How CSI migration will fit into this I'm uncertain of, we really want to try and leave the CSI migration stuff for a different proposal

Reviewer 3: We will have to gate on CSI migration being finished when we do this, @Danil-Grigorev, know that we know more about CSI migration, can we add a detail in here to explain that we will have to wait for it (by which I mean the operators will have to wait for it) before the migration happens?

- `machine-config-operator` restart updates the `kubelet` with `--cloud-provider=external` option. At this point the `CCM` is running out-of-tree

> - `CSI Migration` is finished.

Reviewer 6: what about the time between start and finished? Will we lose any capability of the cluster? E.g. no new volumes anymore?

Prerequisites for downgrade:

- The provider must support `CSI Migration` from in-tree storage. `CSI Migration` must be approved (tests, CI, GA) for the provider upgrade by default.

> - `CSI Migration` should support reverse migration.

Reviewer 6: What are prerequisite for a downgrade? What if these don't apply for the target (=downgrad) version?

## Release Signoff Checklist

- [x] Enhancement is `implementable`
- [x] Design details are appropriately documented from clear requirements
- [x] Test plan is defined
- [x] Graduation criteria for tech preview, GA
- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

## Summary

This enhancement proposal describes the migration of cloud platforms from the [deprecated](https://github.com/kubernetes/kubernetes/blob/222cae36ec667c2f12c8aaa73a86e3cf01d80f7e/cmd/kubelet/app/options/options.go#L389) [in-tree cloud providers](https://v1-18.docs.kubernetes.io/docs/concepts/cluster-administration/cloud-providers/) to [Cloud Controller Manager](https://kubernetes.io/docs/concepts/architecture/cloud-controller/) services that implement the`external cloud provider` [interface](https://github.com/kubernetes/cloud-provider/blob/master/cloud.go).

## Motivation

Using Cloud Controller Managers (CCMs) is the Kubernetes [preferred way](https://kubernetes.io/blog/2019/04/17/the-future-of-cloud-providers-in-kubernetes/) to interact with the underlying cloud platform as it provides more flexibility and freedom for developers.

> It replaces existing in-tree cloud providers, which have been deprecated and will be permanently removed approximately in Kubernetes `v1.23`.

Reviewer 5: Probably it would be worthy to link to upstream document, just to be able to quickly confirm it still holds.

#### Story 3

As an OpenShift developer responsible for the cloud controllers, I want to share more code with upstream Kubernetes in order to ease feature and bug fix contributions in both directions.

#### Story 4

We’d like to discuss technical details related to a specific cloud in a SIG meeting with people who are also involved into development in this domain, and that way gain useful insights into the cloud infrastructure, improve the overall quality of our features, stay on top of the new features, and improve the relations with maintainers outside of our company, which nevertheless share with us a common goal.

### Implementation Details

The `cluster-cloud-controller-manager-operator` implementation will be hosted within its own OpenShift [repository](https://github.com/openshift/cluster-cloud-controller-manager-operator). The operator will manage `CCM` provisioning for supported cloud providers.

#### Operator resource management strategy

Every cloud manages its configuration differently, but shares a common overall structure.

> 1. The operator will be provisioned in `openshift-cloud-controller-manager` namespace.

Reviewer 11: I think you have to prefix it with `-operator`

2. `CCCMO` will provision operands under the `openshift-cloud-controller` namespace.

> 3. The pod running `CCM` controllers in one or more (Azure) containers will be managed by a `Deployment`.

Reviewer 5: Where does CCCMO knows which CCM to deploy?

Author: Current source-of-truth is the `Infrastructure` resource. CCCMO will consume whatever is/will be the equivalent of it in the future and select a collection of templates to apply.

Reviewer 5: Can you point it out explicitly in this document?

- Be built with consideration for potential future implementations of hybrid OpenShift clusters which support multiple cloud providers simultaneously (ie ensure resources created for each provider by the operator do not clash with resources for other providers).

*Note: static resource creation is delegated to CVO (SA, RBAC, Namespaces, etc.)*

#### CVO management

The operator image will be built and included in every release payload.
It is expected to be deployed and running at Kubernetes operators level ([10-29](https://github.com/openshift/cluster-version-operator/blob/34010292c3abd2582b687e9c0ef76c5924998f39/docs/dev/operators.md#how-do-i-get-added-as-a-special-run-level)) right after the `kube-controller-manager-operator` at runlevel [25](https://github.com/openshift/cluster-kube-controller-manager-operator/tree/master/manifests).

Proposed `CCMO` manifests runlevel is `26`.

#### Cloud provider management

Each cloud provider will be hosted under its own repository. Those repositories will be initially bootstrapped by forking upstream ones.

> The `CVO` will be responsible for provisioning static resources, such as `ServiceAccounts`, `Roles/RoleBindings`, `ServiceMonitors` for metrics for the operator. Then, the operator will  be responsible for constructing and creating cloud provider resources.

Reviewer 12: minor nit
```suggestion
The `CVO` will be responsible for provisioning static resources, such as `ServiceAccounts`, `Roles/RoleBindings`, and `ServiceMonitors` for the operator. Then, the operator will  be responsible for constructing and creating cloud provider resources.
```

#### CCM migration from KCM pod

Currently the cloud-specific controller loops are running inside the `kcm` pod. It is ok to run in-tree and out-of-tree `CCM` simultaneously in a cluster for a short amount of time, or have a short amount of downtime between the moment in-tree plugin shuts down, and out-of-tree starts, but this gap should not be experienced by components, which rely on `CCM` functionality. Those are:
- `openshift-ingress` with `Route` resource management: non-operational without `CCM`.
- `kubelet`: can't initialize a new `Node` without `CCM`. Consumes `cloudprovider.Interface` in order to collect `Node` [address](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/nodestatus/setters.go#L65), [populate](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_node_status.go#L388-L435) `spec.ProviderID` and `Labels`, and [collect](https://github.com/kubernetes/kubernetes/blob/master/cmd/kubelet/app/server.go#L960-L978) `Node` name.

> *Note: `storage`: in-tree plugins including storage implementation will remain untouched by the migration. This controller will remain preserved by using the `--external-cloud-volume-plugin` flag on `kcm` pod. Although `CCM` controller loops itself does not require HA setup as they does not store any state, the in-tree storage plugins are, and they will be removed together with the `CCM` from in-tree when it happens.*

Author: It is not there.

*Alternative approach (long-term):*

An upstream effort in this [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cloud-provider/20191004-out-of-tree-credential-providers.md) attempts to remove the `Kubelet` dependency on in-tree credential issuing process, and instead use a similar approach with pluggable credentials providers. This interface is supposed to be moved into the `Alpha` stage with this [PR](https://github.com/kubernetes/kubernetes/pull/94196) in the `1.20` release. Finally, at `1.23` the `kubelet` will no longer [support](https://github.com/kubernetes/enhancements/pull/1665#issuecomment-703140901) in-tree cloud providers.

#### Resource changes

#### Kubelet

In OpenShift, `machine-config-operator` manages `kubelet` configuration.
Based on the [Infrastructure value](https://github.com/openshift/api/blob/master/config/v1/types_infrastructure.go) MCO currently sets the correct `cloud-config` value.

> To support external providers, MCO should not set the `cloud-config` value anymore.

Author: Could I get some links?

Additionally, it must set the value of the `cloud-provider` [flag](https://github.com/openshift/machine-config-operator/blob/2a51e49b0835bbd3ac60baa685299e86777b064f/templates/worker/01-worker-kubelet/_base/units/kubelet.service.yaml#L32) to `external`.
This informs `kubelet` that an out-of-tree provider will finish the `Node` initialization process.

While `kubelet` itself does not directly contact or interact with the `CCM`, formation of a functional `Node` in Kubernetes does rely on the `CCM`.
With in-tree providers, `kubelet` looks up Node IP Addresses, Instance IDs and other metadata directly from the cloud provider and places this information onto the `Node` object.

> When using out-of-tree, the `kubelet` taints the `Node` with `node.cloudprovider.kubernetes.io/uninitialized: true` and does not look up any information.

Author: Yes, they should for completeness. This taint only blocks from assigning a real IP address and surfacing other metadata from the instance where the Node is located (instance type/size/zone)

It is then the responsibility of the `CCM` to look up the Node IP Addresses, Instance IDs and other metadata and labels, place the information onto the `Node` object and then remove the taint from the `Node`.

The information previously set by `kubelet`, now `CCM`, is required to make a `Node` schedulable.
Without this information, the `scheduler` cannot schedule `Pods` that have any sort of affinity that relies on `Node` metadata (eg failure domain).

#### Kube Controller Manager and Kube API Server

In OpenShift, `kube-controller-manager` and `kube-apiserver` are managed by `cluster-kube-controller-manager-operator` and `cluster-kube-apiserver-operator` respectively.
When switching to out-of-tree providers, the `cloud-provider` option [should be removed](https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/#running-cloud-controller-manager) from their deployments.

##### Kube Controller Manager changes to prolong support for in-tree storage

> In the first phase of the transition all cloud providers will still rely on the in-tree storage. `--external-cloud-volume-plugin=<some-provider>` set on `KCM` pod will preserve the in-tree storage controller loops.

Reviewer 12: i find this a little confusing, is this saying that `--external-cloud-volume-plugin=KCM` will preserve the in-tree controller?

Author: --external-cloud-volume-plugin=aws or some other cloud provider. It will preserve in-tree persistent volume controllers. I'll make some changes to the text.

At the later phases, CSI support and CSI migration will be responsible for removing this flag from the `KCM`.

> With the `--external-cloud-volume-plugin` flag set, existing designs for this transition are being preserved. The storage will temporarily become a lone in-tree controller depending on the cloud, and could be excluded by setting this flag off. This means that `kcm` will still initialize the `cloudprovider.Interface`, but this will only be used by storage controllers.

Reviewer 12: i found this paragraph slightly confusing. i think you are saying the following things:

1. the `--external-cloud-volume-plugin` flag will allow us to preserve existing, in-tree, functionality for storage
2. not setting this flag will preserve the in-tree usage

the confusing part to me is this:
> With the `--external-cloud-volume-plugin` flag set, existing designs for this transition are being preserved.

i think the message here is that by having this flag we can preserve the existing in-tree design (eg not using the flag), while enabling the out-of-tree design (eg by setting the flag). is that accurate?

Author: It is slightly different then that. `--external-cloud-volume-plugin=aws` will preserve in-tree PV and PVC controllers hosted in kube-controller-manager, while the `--cloud-provider=external` set *at the same time* will effectively disable some controllers (there is a map of them, don't give precise naming for that reason - [here](https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/controllermanager.go#L432-L437) )yet others, like https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/controllermanager.go#L439 will still use initialized cloud interface and be enabled. Volumes could be attached and detached. 

Missing piece is CSI, which will allow full operation on volumes, like creation provisioning for PVC etc. This flag is set only on KCM, which leads to `kubelet` lacking ~any~ kind of initialization for cloud. Therefor all storage related parts located in `kubelet` will remain non-functional. Until CSI is enabled

Reviewer 12: ok, thanks for that detailed explanation. i think this section should be cleaned up to reflect what you said here, it makes it much clearer imo.

#### Pre-release / Development Feature Gate

> In order to achieve desired cluster configuration for cloud components during development and technical preview, we will add a new OpenShift [FeatureGate](https://docs.openshift.com/container-platform/4.5/nodes/clusters/nodes-cluster-enabling-features.html) resource named `ExternalCloudProvider`, which will allow developers and customers to force their cluster to upgrade from in-tree to out-of-tree cloud providers.

Reviewer 10: "resource named `ExternalCloudProvider` " leaves impression that there will be a Custom Resource (CR) of `kind: FeatureGate` with name `ExternalCloudProvider`, which sounds very odd.

Do you want to say that "external cloud providers will be enabled when `featureSet: TechPreviewNoUpgrade` is used" ? Do you really need to have anything called `ExternalCloudProvider` and what it actually is?

This feature gate will be shipped with `TechPreviewNoUpgrade` `FeatureSet`, preventing upgrades.

The feature gate will be used to determine whether or not operators should migrate their components to the out-of-tree configuration.
If the feature gate is found, the operators for components such as `kube-apiserver`, `kube-controller-manager` and `kubelet` will update their component's configuration accordingly to allow hosting external `CCM` for current platform.

Once an out-of-tree provider implementation is ready to be released (GA), it will no longer depend on the feature gate, allowing production clusters to automatically upgrade to the out-of-tree implementation during 4.N to 4.N+1 upgrade.

Once all providers are moved out-of-tree, the feature gate will be safe to remove.

This feature gate will be used in CI tasks, to test migration stability.

#### Migration procedure

> The process of the migration will be managed by corresponding operators. Communication will be done via `CloudControllerOwner` condition on the operator's `ClusterOperator` resource.

Reviewer 10: Having custom conditions on `ClusterOperator` is quite unusual. Typically, custom conditions are on objects in `operator.openshift.io/v1` group, such as `KubeControllerManager` and you could add `CloudControllerManager` kind for your cccmo.

Author: It is better to use common resource here. Each operator has necessary RBAC permissions, and the mechanism is very similar to the upstream enhancement for CCM migration: https://github.com/kubernetes/enhancements/blob/473d6a094f07e399079d1ce4698d6e52ddcc1567/keps/sig-cloud-provider/20190422-cloud-controller-manager-migration.md They are however still implementing this feature, and they don't intend to automate the process of using this.

```yaml
conditions:
- type: CloudControllerOwner
  status: True
```

`status` will indicate the state of the ownership for this code. Setting it to `False` on one `ClusterOperator` will unblock resource provisioning for the other operator.

The plan for migration is the following:

> 1) `kubelet` and `kube-controller-manager`.

Reviewer 10: Kubelet does not have ClusterOperator instance. Do you mean MCO?

>     1) Set the `CloudControllerOwner` condition on the `ClusterOperator` with `status=True`.

Reviewer 3: Just to make this more explicit
```suggestion
    1) Set the `CloudControllerOwner` condition on their own `ClusterOperator` with `status=True`.
```

    2) Disable cloud loops in its operands.

>     3) Wait for the new resource version rollout to complete and become stable.

Reviewer 10: "rollout to complete and become stable" - does it mean restart of all kubelets? Is there a way how to check that all kubelets were restarted and have required cmdline args? We may need something very similar for CSI migration and we haven't found anything robust enough when multiple MachineConfigs are changed / deployed in parallel.

Author: Whatever sets the `Progressing=False` condition on the ClusterOperator will be responsible to report that. We only need to switch off the previous flags from `KCM`, so at that moment the job is complete.

>     4) Indicate the completion of the migration by setting `status=False` on the `CloudControllerOwner` condition.

Reviewer 10: What if a new node joins the cluster later? Can it run an unconfigured kubelet, i.e. with in-tree cloud provider for some short time? Will something turn Kubelet/MCO `ClusterOperator` to `status=True`?

Author: It can. Initialized kubelet will set the taint on the Node object, which `CCM` will be responsible to remove after. Only MCO can set this condition, and it will be healthy at the moment to report `status=True`.

2) `cloud-controller-manager`

>     1) Set the `CloudControllerOwner` condition on the `ClusterOperator` with `status=False`.

Reviewer 3: ```suggestion
    1) Set the `CloudControllerOwner` condition on its `ClusterOperator` with `status=False`.
```

    2) Wait for preceding operators to have the `CloudControllerOwner` condition with `status=False`.
    3) Provision out-of-tree resources.
    4) Indicate the ownership transfer by setting `status=True` on the `CloudControllerOwner` condition.
    3) Wait for the resources rollout to finish.

> Once all resources are up and running, the migration is completed.

Reviewer 10: There should be a short note somewhere about default values - when running an older kubelet or controller-manager that does not know anything about external cloud providers yet, missing `CloudControllerOwner` condition should tell the world around (CCCMO?) that in-tree cloud provider is used there.

Once the provider is moved to out-of-tree, the migration mechanism will be disabled and the helper `CloudControllerOwner` condition will be removed. When all existing in-tree providers are moved to out-of-tree, the implementation will be safe to remove.

#### Bootstrap changes

Once an out-of-tree provider is released (GA), the `CCM` will be created as a static pod on the bootstrap node, to ensure swift removal of the `node.cloudprovider.kubernetes.io/uninitialized: true` taint from any newly created `Nodes`. Later stages, including cluster upgrades will be managed by an operator, which will ensure stability of the configuration, and will run the `CCM` in a `Deployment`.
Initial usage of the static pod is justified by the need for `CCM` to initialise `Nodes` before the `kube-scheduler` is able to schedule regular workloads (eg the operator for `CCM`).

The order will be the following:

1) Start `etcd` and `kube-api-server`.
2) Run the `CCM`, which will be provisioned as a static pod.

> 3) Start the `kube-scheduler`.

Reviewer 3: Based on Stefan's comments here, we probably don't need this ordering list at all here

> `CCCMO` will provide a way to generate initial manifests that allow deployment of `CCM` as a static pod on the bootstrap machine.

Reviewer 3: Based on Stefan's comment
```suggestion
`CCCMO` will provide a render command to generate initial manifests that allow deployment of `CCM` as a static pod on the bootstrap machine. This will follow the style of other operators (eg `KCMO`).
```

*Note: metrics state is a subject to change, and we will have influence upon adding back ones required by OpenShift before transition to out-of-tree*

### Disaster Recovery

In the original `KCM` implementation, where the `CCM` is currently hosted, is using a [semi-automated](https://docs.openshift.com/container-platform/4.5/backup_and_restore/disaster_recovery/scenario-2-restoring-cluster-state.html) approach to disaster recovery, introducing the `forceRedeploymentReason` option in the managed CR, triggering the redeployment of the static pods, running the `KCM` code.

`CCM` approach is simpler, as we run the `CCCMO` which manages the `Deployment`. Re-populating etcd will revert the old image, and the previous version of the `CCCMO` code will re-apply and restart the old `CCM` configuration.

### Upgrade/Downgrade strategy

A set of `cloud-controller-manager` pods will be running under a `Deployment` resource, provisioned by the operator. The leader-election will preserve the leadership during the cluster updates over cloud-specific loops.

Upgrades from previous versions of OpenShift will look like:

> #### Post 4.7 (in-tree to out-of-tree per provider)

Reviewer 3: This slipped
```suggestion
#### Post 4.8 (in-tree to out-of-tree per provider)
```

Each provider will be released independently as and when the providers are deemed stable enough.

Upgrade:

- New `cluster-kube-apiserver-operator` and `cluster-kube-controller-manager-operator` version updates the `kube-apiserver` and the `KCM` and removes the `--cloud-provider` flag. `cluster-kube-controller-manager-operator` in addition sets the `--external-cloud-volume-plugin=<some-platform>` flag on the `KCM` pods to ensure in-tree storage preservation.
- `cluster-cloud-controller-manager-operator` starts and creates out-of-tree `CCM` resources for the provider.
- `machine-config-operator` restart updates the `kubelet` with `--cloud-provider=external` option. At this point the `CCM` is running out-of-tree

Downgrade:

- `cluster-cloud-controller-manager-operator` de-provisions CCM resources for any platform.

> - `cluster-kube-apiserver-operator` and `cluster-kube-controller-manager-operator` restart updates the `kube-apiserver` and the `kube-controller-manager` with the `--cloud-provider=<some-platform>` option.

Reviewer 10: Is order here important? Downgrade updates operators in the same order as upgrade, i.e. api server first, then kube-controller-manager, then the rest.

I am not sure if CVO is going to remove cccmo on downgrade - it may not be present in the older release manifests.

Author: No, the order is not important. We merge a code located in two places into a third one. Backwards copy pasting it into two places and removing from the third. CCCMO should only deprovision it's resources. It can hang for a while.

- `cluster-kube-controller-manager-operator` removes the `--external-cloud-volume-plugin=<some-platform>` flag from the `KCM` pods as `--cloud-provider=<some-platform>` superseeds it.
- `machine-config-operator` restart applies an old machine configuration, and cause `kubelet` to restart with `--cloud-provider=<some-platform>` option.

#### Providers, already running out-of-tree CCM (4.8+)

Upgrade:

- `cluster-cloud-controller-manager-operator` updates the state of resources for any in-tree provider defaulting to this since previous releases.

Downgrade:

- `cluster-cloud-controller-manager-operator` downgrades the state of resources for any in-tree provider defaulting to this since previous releases.

> ### Version Skew Strategy

Reviewer 10: Current eus->eus upgrade strategy expectes that some kubelet updates are skipped. I.e, kcm may be 4.10, while kubelet is 4.8 (numbers are purely illustrative). Can it break anything, esp. when the feature is GA in 4.10 and still optional / disabled by default in 4.8?

Author: Not unless we integrate migration functionality in 4.8. That's why starting earlier is important. The gap may result in core providers not reporting necessary conditions in time, so the migration may not initiate.

Reviewer 3: Is this still true? I was of the impression we had decided EUS would mean 4 baby upgrades to get between EUS versions? IE we can't skip ahead?

Actions:

- Make sure the Cinder CSI driver is supported in OpenShift.
- Make sure the Cinder CSI migration is supported in OpenShift.
- `KCM` excludes the rest of cloud-controller loops by running CSI storage out-of-tree.

#### Examples for other cloud-controller-manager configurations

- AWS: https://github.com/kubernetes/cloud-provider-aws/. **Demo:** , to reproduce see: https://github.com/cloud-team-poc/cloud-provider-aws. [![asciicast](https://asciinema.org/a/9QSHP9ZRWvwWNoxEW2rEhLm5m.svg)](https://asciinema.org/a/9QSHP9ZRWvwWNoxEW2rEhLm5m)
- Azure: https://github.com/kubernetes-sigs/cloud-provider-azure
Azure separated the `CCM` and the "cloud-node-manager" (`CNM`) into two binaries: https://github.com/kubernetes-sigs/cloud-provider-azure/tree/master/cmd.
- GCP: https://github.com/kubernetes/cloud-provider-gcp
- vSphere: https://github.com/kubernetes/cloud-provider-vsphere
vSphere introduced an additional service to expose internal API server: https://github.com/kubernetes/cloud-provider-vsphere/blob/master/manifests/controller-manager/vsphere-cloud-controller-manager-ds.yaml#L64

### Risks and Mitigations

#### Specific CCM doesn’t work properly on OpenShift

> The `CCMs` have not been tested on OCP yet. Unknown issues may arise.

Reviewer 13: With a sev/liklihood of MedHi and Med respectively, what is the mitigation plan around CCM's now working properly on OpenShift?

Reviewer 3: I wonder if the likelihood is actually too high for this risk? For some of the providers the code is already running in OpenShift and is literally just being moved from one binary to another binary. That said, Openstack is a complete new implementation IIRC, though, it's working in vanilla K8s clusters to replace the existing in tree provider so I would imagine it's relatively stable and functional at this point.

We will of course be testing the providers thoroughly using the origin E2E suite to ensure compatibility.

> as this will cause a lot of churn in the MCO repo during a period of heavy development.

@kikisdeliveryservice Could you please expand on what you mean by this? My expectation was that the changes to the code in MCO as we add new providers would be a PR to adjust the platform switch statement to return external for certain providers if the featuregate is present? We should be able to raise PRs to each component and run tests to validate that the functionality is sufficient for the external provider before merging, I'm not sure I would classify this as churn?

Author: @kikisdeliveryservice The reason for changes in MCO is based on intent to decrease this likelihood. It might mean we need to do some adjustments to bootstrap procedure for MCO at some moment to properly test it, but this is still under discussion. Other changes to MCO are exactly as Joel described. We are mostly planning to remove things in loner term.

  - `kubelet` dependency requires removal of taints from newly created Nodes. Without CCM, new nodes will not become schedulable, existing nodes are unaffected.
  - CCMs manage `LoadBalancer` `Service` resources in all in-tree cloud implementations, except vSphere, so their creation will not be possible while the CCM is down. This will affect the [ingress-operator](https://github.com/openshift/cluster-ingress-operator). Cloud networking across `Nodes` is [disabled](https://github.com/openshift/cluster-kube-controller-manager-operator/blob/8be94db9da523152af2268bd7f891fe089a424eb/bindata/v4.1.0/config/defaultconfig.yaml#L8-L9) by default for all clouds, so the routes will not be provisioned for the Node by the `CCM` (which is done by `Routes()` method in the interface), and so not required for SDN functionality.
  - Storage CSI migration will be required for all cloud providers.

Q: Should we reuse the existing cloud provider config or generate a new one?
A: CCM config is backward compatible with the in-tree cloud provider, therefore we can reuse it.

Q: How to migrate PVs created by the in-tree cloud provider?

> A: [CSIMigration](https://kubernetes.io/blog/2019/12/09/kubernetes-1-17-feature-csi-migration-beta/) is the preferred option. It graduates in GA in 1.19 (1.20).

Reviewer 10: It will graduate in 1.22 at the earliest.

Q: What are the thoughts about certificate management?
A: At present, only the GCP provider has any CSR approval mechanism built in. This functionality is used to approve `NodeClient` and `NodeServer` certificates for Nodes on GCP.
This functionality should not be required as OpenShift handles certificate approval via the `cluster-machine-approver` component.

Q: Does every node need a CCM?
A: No. The cluster only needs one active `CCM` at any time. A `Deployment` will manage the `CCM` pod and will have 2 replicas which will use leader election to nominate an active leader. [Source](https://kubernetes.io/docs/concepts/overview/components/#cloud-controller-manager)

Q: How metrics are affected by the CCM migration?
A: Currently the `CCM` implementation is using the same set of metrics from the core repository, all existing metrics will still be available post migration. The `CCMO` will be required to create a `ServiceMonitor` to instruct prometheus to collect metrics from the `CCM` pod.

> 

Reviewer 5: Other questions which not necessarily needs answers right away are:
1. Do we plan to run multiple CCMs for different clouds in parallel?

Author: Yes. This is what GCP does in their solutions according to what I heard. It is possible to do that, only correct naming and leader separation is needed.

Reviewer 3: This question is answered in the non-goals section, we do not plan to support multiple clouds from a single cluster, so no multiple CCMs

Author: Agree with Joel. No, not until the need for this feature is communicated. It is possible

Reviewer 5: Right, apologies I missed it.

The `cloud-provider` and `cloud-config` flags from `kubelet` will be removed in 1.23 - [link](https://github.com/kubernetes/kubernetes/pull/90408).

## Timeline

### Upstream

Follow the Kubernetes community discussions and implementation/releases

- vSphere support has graduated to beta in [v1.18](https://github.com/kubernetes/enhancements/issues/670) (Openshift 4.5)

- Azure support goes into beta in [v1.20](https://github.com/kubernetes/enhancements/issues/667) (Openshift 4.7)

### OpenShift

#### OCP 4.8

- AWS on out-of-tree `CCM` tech-preview. `CSI` is optional (via OLM) since [4.6](https://github.com/openshift/enhancements/blob/master/enhancements/storage/csi-driver-install.md#ocp-45-kubernetes-118).
- OpenStack platform on out-of-tree `CCM` graduates as tech-preview.

#### OCP 4.9

- OpenStack upgrade and installation with the out-of-tree CCM by default in 4.9. (GA)

#### OCP 4.10

- vSphere, `CSI` release plan for tech-preview is 4.7+.
- Azure.

> - GCP.

Reviewer 10: Regarding the timeline - CCM GA must go in sync with CSI migration GA, as CCM implies CSI migration and
we can't have in-tree volume plugin (GA) replaced by CSI driver which is beta / tech preview.

From upstream sig-storage plan, which changes every 3 weeks:

* AWS, GCE, Azure: GA in 1.23 (4.10)
* OpenStack: GA in 1.22 or 1.23 (4.9/4.10)
* vSphere: GA in 1.24 (4.11)

The migration is reaching beta in various releases and from storage team point of view we can do tech preview in 4.8 with GCE, OpenStack and AWS (all of them or a subset, whatever suits you).

Reviewer 10: Update after Kubernetes 1.21 code freeze:

* AWS, GCE, OpenStack, Azure: GA in 1.23 (4.10)
* vSphere: GA in 1.24 (4.11)

From CSI point of view, in 4.8 we need tech preview of OpenStack CSI + CCM migration, AWS is highly recommended (to get feedback of most customers). In 4.9, we may need tech preview of everything except vSphere, to catch upstream GA in 4.10.

#### Story 4

We’d like to discuss technical details related to a specific cloud in a SIG meeting with people who are also involved into development in this domain,
and that way gain useful insights into the cloud infrastructure, improve the overall quality of our features, stay on top of the new features, and improve the relations with maintainers outside of our company, which nevertheless share with us a common goal.

### Implementation Details

The `cluster-cloud-controller-manager-operator` implementation will be hosted within its own OpenShift [repository](https://github.com/openshift/cluster-cloud-controller-manager-operator). The operator will manage `CCM` provisioning for supported cloud providers.

#### Operator resource management strategy

Every cloud manages its configuration differently, but shares a common overall structure.

1. The operator will be provisioned in `openshift-cloud-controller-manager-operator` namespace.
2. `CCCMO` will provision operands under the `openshift-cloud-controller-manager` namespace.
3. The pod running `CCM` controllers in one or more (Azure) containers will be managed by a `Deployment`.

> 4. The pods will be expected to run on the control-plane `Nodes` using `hostNetwork`.

Author: @rtheis This actually came up recently, and now upon approach @JoelSpeed tested `hostNetwork` is required for the `cloud-node` controller which initialise master Nodes during cluster installation. Is this still an issue for you? Enhancement will be updated shortly.

> 4. The pods will be expected to run on the control-plane `Nodes` using `hostNetwork`.

Reviewer 14: Running on the host network is not a functional problem for the IBM CCM.  We would prefer to run on the pod network if possible.  This is not an issue for us.  Thanks for checking.

*[Azure](https://github.com/kubernetes-sigs/cloud-provider-azure/tree/master/cmd) is running them as separate binaries, other known providers keep them together.*
*The rest of the controller loops are provider dependent features, built under their consideration. An example is the [GCP](https://github.com/kubernetes/cloud-provider-gcp/blob/08a120e3a8936c200ef77d0e3384eb730ad6be77/cmd/gcp-controller-manager/loops.go#L45) certificate approval and node annotator controllers.*
*In OpenShift the [cluster-machine-approver](https://github.com/openshift/cluster-machine-approver) will take any additionally required responsibility to auto-approve `CSR`*

`CCCMO` will:

- Create and manage the cloud-specific `Deployment`, based on the selected `platform` in the `Infrastructure` resource.
  - This will run the `CCM` pods on the `control-plane` nodes, and will tolerate `cloudprovider.kubernetes.io/uninitialized` and `node-role.kubernetes.io/master` taints.

>   - This will run at least `2` replicas of the pods with anti-affinity settings to prevent scheduling on the same machine to allow some fault-tolerance if a machine were to fail.

Reviewer 3: OpenShift should always deploy master machines to different zones, by restricting the CCM pods to running on master nodes and scheduling with anti-affinity to each other, they should be spread across failure domains without having to rely on cloud provider information for scheduling of the CCM pod

Proposed `CCMO` manifests runlevel is `26`.

#### Cloud provider management

Each cloud provider will be hosted under its own repository. Those repositories will be initially bootstrapped by forking upstream ones.

The `CVO` will be responsible for provisioning static resources, such as `ServiceAccounts`, `Roles/RoleBindings`, and `ServiceMonitors` for the operator. Then, the operator will  be responsible for constructing and creating cloud provider resources.

Forked provider repositories will also contain cloud-provider specific code, and a set of binaries for each of them to build and run. Each provider will be built inside a single image, which will be included in the release payload. The responsibility for the operator will be to choose which provider image should be used in a cluster at any moment.

#### CCM migration from KCM pod

Currently the cloud-specific controller loops are running inside the `kcm` pod.

> It is ok to run in-tree and out-of-tree `CCM` simultaneously in a cluster for a short amount of time, or have a short amount of downtime between the moment in-tree plugin shuts down, and out-of-tree starts, but this gap should not be experienced by components, which rely on `CCM` functionality.

Reviewer 3: This was always our plan, CCCMO should be gated on a status condition that is set by KCMO to show that it has configured the KCMs to use external cloud providers. This should be explained somewhere within this doc

#### Examples for other cloud-controller-manager configurations

- AWS: https://github.com/kubernetes/cloud-provider-aws/. **Demo:** , to reproduce see: https://github.com/cloud-team-poc/cloud-provider-aws. [![asciicast](https://asciinema.org/a/9QSHP9ZRWvwWNoxEW2rEhLm5m.svg)](https://asciinema.org/a/9QSHP9ZRWvwWNoxEW2rEhLm5m)
- Azure: https://github.com/kubernetes-sigs/cloud-provider-azure
Azure separated the `CCM` and the "cloud-node-manager" (`CNM`) into two binaries: https://github.com/kubernetes-sigs/cloud-provider-azure/tree/master/cmd.
- GCP: https://github.com/kubernetes/cloud-provider-gcp
- vSphere: https://github.com/kubernetes/cloud-provider-vsphere
vSphere introduced an additional service to expose internal API server: https://github.com/kubernetes/cloud-provider-vsphere/blob/master/manifests/controller-manager/vsphere-cloud-controller-manager-ds.yaml#L64

### Risks and Mitigations

#### Specific CCM doesn’t work properly on OpenShift

The `CCMs` have not been tested on OCP yet. Unknown issues may arise.

Severity: medium-high

Likelihood: medium

The likelihood of the risk above varies per provider. None of the providers will be shipped without testing.

### Test Plan

> **All following test plan details have a hard requirement - support for teach preview FeatureGate described in this [section](#Pre-release-/-Development-Feature-Gate). No development or testing could be done before the functionality is implemented.**

Reviewer 14: Adding comment here but it applies throughout, for cloud providers that were never in-tree (e.g. IBM cloud provider), can we have a streamline set of requirements that go directly to the CCM implementation only?

Q: Who talks to CCM?

- A: CCM exposes health, readiness, metrics and config endpoints similar to most core Kubernetes controllers.
These endpoints are accessed by the process host's Kubelet to ensure the process is running smoothly and Prometheus to gather metrics.
No other components talk to the CCM directly. The CCM coordinates with other components by updating resources using the KAS similar to other controllers.

Q: CCM relation to other openshift components, such as SDN and storage? How a non-operational CCM will affect cluster health, which components will take a hit?

- A: A couple of components rely on CCM at different time of the cluster lifetime:
  - `kubelet` dependency requires removal of taints from newly created Nodes. Without CCM, new nodes will not become schedulable, existing nodes are unaffected.
  - CCMs manage `LoadBalancer` `Service` resources in all in-tree cloud implementations, except vSphere, so their creation will not be possible while the CCM is down.
  This will affect the [ingress-operator](https://github.com/openshift/cluster-ingress-operator).

>   Cloud networking across `Nodes` is [disabled](https://github.com/openshift/cluster-kube-controller-manager-operator/blob/8be94db9da523152af2268bd7f891fe089a424eb/bindata/v4.1.0/config/defaultconfig.yaml#L8-L9) by default for all clouds, so the routes will not be provisioned for the Node by the `CCM` (which is done by `Routes()` method in the interface), and so not required for SDN functionality.

Reviewer 14: CCM is not required to implement routes if the CNI component (e.g. Calico) does it.  Will this use case be supported?

Author: According to the cloud interface implementation you could enable/disable parts of it. If your CCM works in such configuration from your testing, then it should work correctly in Openshift as well. But this question should be first discussed with networking team, we can't possibly know all details now.

*Bootstrap phase:*

- Initial configuration with static pods is expected to be using `ConfigMap` credentials, similar to what is [done](https://github.com/openshift/installer/blob/ba6d7fe087eada5a4fc260064c85a484a8c45aaf/data/data/bootstrap/files/usr/local/bin/bootkube.sh.template#L162) in the `kcm` operator.
- Once it is supported, the static pods will also use the `CredentialsRequest` to get cloud credentials.

*Post-install phase:*

- Using a `CredentialsRequest` is the default option, every supported `CCM` within the `Deployment` will request its own set of credentials.
- A set of `CredentialsRequest` resources will be hosted under `CCCMO` repository, and created by the `CVO`, similarly to the [machine-api](https://github.com/openshift/machine-api-operator/blob/6f629682b791a6f4992b78218bfc6e41a32abbe9/install/0000_30_machine-api-operator_00_credentials-request.yaml) approach.

Depending on the cloud-provider, when it is supported, credentials management is deligated to the instance metadata endpoint to be acquired from. `AWS` and `Azure` are currently known to support this feature in some extent.

#### Resource changes

#### Kubelet

> In OpenShift, `machine-config-operator` manages `kubelet` configuration.

Reviewer 15: Some verbiage regarding `windows-machine-config-operator` managing `kubelet` configuration for Windows nodes should be added here.

Author: Added a note that `kubelet` behaviour changes and effects would apply for `WMCO` as well.

Q: Does CCM provide an API? How is it hosted? HTTPS endpoint? Through load balancer?

- A: Not at the moment. The CCM vendors part of the code serving cloud features from the [cloud-provider](https://github.com/kubernetes/cloud-provider). CCM provides a secure TCP connection on port [10258](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cloud-provider/ports.go#L22) and it is only used for  metrics and readiness/health endpoint.

Q: What happens if the kcm leader has no access to ccm? Can it continue its work? Will it give up leadership?

- A: KCM and CCM perform independently after the initial migration process. KCM will be unaffected by CCM being down after the initial migration. The CCM cluster operator will be responsible for transitioning to Unavailable or Degraded in a similar situation.

Q: What are the thoughts about certificate management?

- A: At present, only the GCP provider has any CSR approval mechanism built in. This functionality is used to approve `NodeClient` and `NodeServer` certificates for Nodes on GCP.
This functionality should not be required as OpenShift handles certificate approval via the `cluster-machine-approver` component.

> Q: Does every node need a CCM?

Reviewer 15: I don't see any mention of `cloud-node-controller` in the enhancement. Isn't that also needed for this transition?

Reviewer 3: The `cloud-node-manager` (I think that's the same) is typically included in the CCM. AFAIK only Azure runs it as a separate DaemonSet (certainly for the platforms OCP supports today). I believe that is called out somewhere within this enhancement already.

Looking now it seems the note we had about running the DaemonSet for azure has gone, will make sure we add something back

Author: Added a minor comment describing `CNM` role in this question.

To ensure that cluster disaster recovery procedures can still operate smoothly, we will ensure that core control plane components and their operators tolerate the uninitialized taint, to prevent `CCM` blocking new control plane hosts being added if `CCM` is non-functional.
This will include, but is not limited to: Kube Controller Manager, Etcd, Kube API Server, Networking, Cluster Machine Approver.

#### Windows Machine Config Operator

##### Kubelet Changes

To generate the configuration for Windows nodes, `WMCO` reads the output of MCO's rendered configuration as a basis for its own config for Kubelet on the Windows node.
As we are making changes to the output of the Kubelet service in `MCO` (namely change the value of the `cloud-provider` flag), we will need to verify that `WMCO` reads this flag and copies its value to the Kubelet Windows service.

##### Node Initialization

On most platforms, `Node` initialization is handled centrally by the `CCM`, specifically the Cloud Node Manager (`CNM`) running within it.
However, On certain platforms (e.g. Azure), the `CNM` must be run on the `Node` itself, typically via a `DaemonSet`.

> Since Red Hat cannot supply or support Windows container images, we cannot run a `DaemonSet` for the `CNM` targeted at Windows Nodes as we would do on Linux Nodes.

Reviewer 16: If WMCO is managing CNM, its worth noting that WMCO is managed by OLM and not CVO, so Windows nodes will updated independently of the rest of the cluster.
This means there will be version skew for some period of time, especially if WMCO upgrades are set to be manually approved by the cluster administrator.
I'm not sure if this would cause an issue, but worth keeping in mind.

Author: Added a detail there.

This feature gate will be shipped with `TechPreviewNoUpgrade` `FeatureSet`, preventing upgrades.

The feature gate will be used to determine whether or not operators should migrate their components to the out-of-tree configuration.
If the feature gate is found, the operators for components such as `kube-apiserver`, `kube-controller-manager` and `kubelet` will update their component's configuration accordingly to allow hosting external `CCM` for current platform.

Once an out-of-tree provider implementation is ready to be released (GA), it will no longer depend on the feature gate, allowing production clusters to automatically upgrade to the out-of-tree implementation during 4.N to 4.N+1 upgrade.

Once all providers are moved out-of-tree, the feature gate will be safe to remove.

This feature gate will be used in CI tasks, to test migration stability.

> 

Reviewer 6: Describe what happens when the gate is unset again.

Reviewer 3: General question, what happens with other feature gates when they are unset? Is it expected that they disable the feature, or, given these are experimental and block upgrades, are we supposed to ignore it being unset?

Author: Disable the feature is the answer. Covered in the proposal.

#### Migration procedure

> The process of the migration will be managed by corresponding operators. Communication will be done via conditions on the `ClusterOperator` resource.

Reviewer 5: operator's `ClusterOperator` resource. 

So that it's clear that each operator exposes that information through his own CO.status.

The plan for migration is the following:

1) `kubelet` and `kube-controller-manager`.

>     1) Disable cloud loops in it's operands.

Reviewer 6: nit: its

>     2) Wait for migration to complete (deployment roll out, kubelet restarted)

Reviewer 6: how? Waiting for which signal exactly?

Reviewer 3: This will probably be complex to dive into, but at present, do we know if the KCMO waits until the updated deployment/pods are running/ready before setting it's operator status back to normal? (I assume during a rollout it sets it to progressing but this assumption may not be true). If it does do this, we can piggy back on that. I'm hoping there's also something similar for the kubelet rollouts but we need to check on both of these fronts

Author: I suggest we would wait for `Progressing` condition to be set on false. Similarly, backwards migration will watch for `CCCMO` cluster operator to get into `Progressing` false before starting rollback.

Reviewer 3: I think the operators that are doing any rollout/rollback should only care about the `ExternalCloudProvider` condition and not the progressing condition. I think this question is more a case of when will that be added by the operator.

So for KCMO, we want it to add it only when it has updated the deployment for KCM, and the old pods have all stopped, and the new pods are ready. I'm hoping it has logic for this already but we should double check (and for the Kubelet rollouts too, I imagine MCO has some status it sets on the operator while it's rolling out, that is the logic we should be able to piggy back on)

>     3) Communicate the completion of the migration stage on their `ClusterOperator` resource for the`CCCMO` to pick up.

Reviewer 6: add " via a condition (name to be defined)"

2) `cloud-controller-manager`
    1) Waits for preceeding operators to report `ExternalCloudProvider` condition.
    2) Starts provisioning out-of-tree resources.
3) Once necessary deployments are up and running, the migration is completed.

*Note:`kube-apiserver` migration is not important, as it currently depends on `cloud.Interface` only in establishing SSH node tunelling functionality unused in `OpenShift` and [deprecated](https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-apiserver/app/options/options.go#L197-L205) since 1.9*

Before it is desided to make out-of-tree a default selection for the relase, the 1st step will be initiated by the `FeatureGate` resource described above. In the designated release for switch from in-tree for a particular platform, the 1st step will be a part of default upgrade procedure.

Once the provider is moved to out-of-tree, the migration mechanism will be disabled. When all exsisting in-tree providers are moved to out-of-tree, the implementation will be safe to remove.

> 

Reviewer 6: Will there be a migration away from external to internal?

#### OCP 4.9

##### Technical preview - 4.9

- AWS on out-of-tree `CCM` tech-preview. `CSI` is optional (via OLM) since [4.6](https://github.com/openshift/enhancements/blob/master/enhancements/storage/csi-driver-install.md#ocp-45-kubernetes-118).
- Azure
- OpenStack

##### General availability - 4.9

- Azure Stack. Bootstrap and post-install phase supported by operator.

#### OCP 4.10

##### Technical preview - 4.10

- vSphere
- GCP

##### General availability - 4.10

> - OpenStack upgrade and installation with the out-of-tree CCM by default. (GA)

Reviewer 3: Do we need AWS and Azure in this list too, with the above description as well?

Author: Yep, added a general direction and missed AWS and Azure in 4.10. Does it now look better? @JoelSpeed
